<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R for Mass Spectrometry - 3&nbsp; Identification data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/30-quant.html" rel="next">
<link href="../pages/raw-ms-data.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/20-id.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">R for Mass Spectrometry</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/R4MS/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/R4MS/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/raw-ms-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Raw MS data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20-id.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30-quant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95-annex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/99-si.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Additional materials and help</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#identification-data.frame" id="toc-identification-data.frame" class="nav-link active" data-scroll-target="#identification-data.frame"><span class="header-section-number">3.1</span> Identification data.frame</a></li>
  <li><a href="#keeping-all-matches" id="toc-keeping-all-matches" class="nav-link" data-scroll-target="#keeping-all-matches"><span class="header-section-number">3.2</span> Keeping all matches</a></li>
  <li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data"><span class="header-section-number">3.3</span> Filtering data</a></li>
  <li><a href="#adding-identification-data-to-raw-data" id="toc-adding-identification-data-to-raw-data" class="nav-link" data-scroll-target="#adding-identification-data-to-raw-data"><span class="header-section-number">3.4</span> Adding identification data to raw data</a></li>
  <li><a href="#an-identification-annotated-chromatogram" id="toc-an-identification-annotated-chromatogram" class="nav-link" data-scroll-target="#an-identification-annotated-chromatogram"><span class="header-section-number">3.5</span> An identification-annotated chromatogram</a></li>
  <li><a href="#visualising-peptide-spectrum-matches" id="toc-visualising-peptide-spectrum-matches" class="nav-link" data-scroll-target="#visualising-peptide-spectrum-matches"><span class="header-section-number">3.6</span> Visualising peptide-spectrum matches</a></li>
  <li><a href="#comparing-spectra" id="toc-comparing-spectra" class="nav-link" data-scroll-target="#comparing-spectra"><span class="header-section-number">3.7</span> Comparing spectra</a></li>
  <li><a href="#summary-exercise" id="toc-summary-exercise" class="nav-link" data-scroll-target="#summary-exercise"><span class="header-section-number">3.8</span> Summary exercise</a></li>
  <li>
<a href="#exploration-and-assessment-of-identifications-using-msnid" id="toc-exploration-and-assessment-of-identifications-using-msnid" class="nav-link" data-scroll-target="#exploration-and-assessment-of-identifications-using-msnid"><span class="header-section-number">3.9</span> Exploration and Assessment of Identifications using <code>MSnID</code></a>
  <ul class="collapse">
<li><a href="#step-by-step-work-flow" id="toc-step-by-step-work-flow" class="nav-link" data-scroll-target="#step-by-step-work-flow"><span class="header-section-number">3.9.1</span> Step-by-step work-flow</a></li>
  <li><a href="#analysis-of-peptide-sequences" id="toc-analysis-of-peptide-sequences" class="nav-link" data-scroll-target="#analysis-of-peptide-sequences"><span class="header-section-number">3.9.2</span> Analysis of peptide sequences</a></li>
  <li><a href="#trimming-the-data" id="toc-trimming-the-data" class="nav-link" data-scroll-target="#trimming-the-data"><span class="header-section-number">3.9.3</span> Trimming the data</a></li>
  <li><a href="#parent-ion-mass-errors" id="toc-parent-ion-mass-errors" class="nav-link" data-scroll-target="#parent-ion-mass-errors"><span class="header-section-number">3.9.4</span> Parent ion mass errors</a></li>
  <li><a href="#filtering-criteria" id="toc-filtering-criteria" class="nav-link" data-scroll-target="#filtering-criteria"><span class="header-section-number">3.9.5</span> Filtering criteria</a></li>
  <li><a href="#setting-filters" id="toc-setting-filters" class="nav-link" data-scroll-target="#setting-filters"><span class="header-section-number">3.9.6</span> Setting filters</a></li>
  <li><a href="#filter-optimisation" id="toc-filter-optimisation" class="nav-link" data-scroll-target="#filter-optimisation"><span class="header-section-number">3.9.7</span> Filter optimisation</a></li>
  <li><a href="#export-msnid-data" id="toc-export-msnid-data" class="nav-link" data-scroll-target="#export-msnid-data"><span class="header-section-number">3.9.8</span> Export <code>MSnID</code> data</a></li>
  </ul>
</li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-id" class="quarto-section-identifier"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>Peptide identification is performed using third-party software - there is no package to run these searches directly in R. When using line search engines it possible to hard-code or automatically generate the search command lines and run them from R using a <code><a href="https://rdrr.io/r/base/system.html">system()</a></code> call. This allows to generate these reproducibly (especially useful if many command lines need to be run) and to keep a record in the R script of the exact command.</p>
<p>The example below illustrates this for 3 mzML files to be searched using <code>MSGFplus</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mzmls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"file_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="st">".mzML"</span><span class="op">)</span></span>
<span><span class="va">mzids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"mzML"</span>, <span class="st">"mzid"</span>, <span class="va">mzmls</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cmds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"java -jar /path/to/MSGFPlus.jar"</span>,</span>
<span>       <span class="st">" -s "</span>, <span class="va">mzmls</span>,</span>
<span>       <span class="st">" -o "</span>, <span class="va">mzids</span>,</span>
<span>       <span class="st">" -d uniprot.fas"</span>,</span>
<span>       <span class="st">" -t 20ppm"</span>,</span>
<span>       <span class="st">" -m 0"</span>,</span>
<span>       <span class="st">" int 1"</span><span class="op">)</span></span>
<span><span class="va">cmds</span></span>
<span><span class="co">##  [1] "java -jar /path/to/MSGFPlus.jar -s file_1.mzML -o file_1.mzid -d uniprot.fas -t 20ppm -m 0 int 1"</span></span>
<span><span class="co">##  [2] "java -jar /path/to/MSGFPlus.jar -s file_2.mzML -o file_2.mzid -d uniprot.fas -t 20ppm -m 0 int 1"</span></span>
<span><span class="co">##  [3] "java -jar /path/to/MSGFPlus.jar -s file_3.mzML -o file_3.mzid -d uniprot.fas -t 20ppm -m 0 int 1"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="identification-data.frame" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="identification-data.frame">
<span class="header-section-number">3.1</span> Identification data.frame</h2>
<p>Let’s use the identification from <code>msdata</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">idf</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">ident</a></span><span class="op">(</span>full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">idf</span><span class="op">)</span></span>
<span><span class="co">##  [1] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The easiest way to read identification data in <code>mzIdentML</code> (often abbreviated with <code>mzid</code>) into R is to read it with the <code><a href="https://rdrr.io/pkg/PSMatch/man/PSM.html">readPSMs()</a></code> function from the <a href="https://rformassspectrometry.github.io/PSMatch/"><code>PSMatch</code></a> package<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. The function will parse the file and return a <code>DataFrame</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/PSM">PSMatch</a></span><span class="op">)</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/PSM.html">PSM</a></span><span class="op">(</span><span class="va">idf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">##  [1] 5802   35</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">##   [1] "sequence"                 "spectrumID"              </span></span>
<span><span class="co">##   [3] "chargeState"              "rank"                    </span></span>
<span><span class="co">##   [5] "passThreshold"            "experimentalMassToCharge"</span></span>
<span><span class="co">##   [7] "calculatedMassToCharge"   "peptideRef"              </span></span>
<span><span class="co">##   [9] "modNum"                   "isDecoy"                 </span></span>
<span><span class="co">##  [11] "post"                     "pre"                     </span></span>
<span><span class="co">##  [13] "start"                    "end"                     </span></span>
<span><span class="co">##  [15] "DatabaseAccess"           "DBseqLength"             </span></span>
<span><span class="co">##  [17] "DatabaseSeq"              "DatabaseDescription"     </span></span>
<span><span class="co">##  [19] "scan.number.s."           "acquisitionNum"          </span></span>
<span><span class="co">##  [21] "spectrumFile"             "idFile"                  </span></span>
<span><span class="co">##  [23] "MS.GF.RawScore"           "MS.GF.DeNovoScore"       </span></span>
<span><span class="co">##  [25] "MS.GF.SpecEValue"         "MS.GF.EValue"            </span></span>
<span><span class="co">##  [27] "MS.GF.QValue"             "MS.GF.PepQValue"         </span></span>
<span><span class="co">##  [29] "modPeptideRef"            "modName"                 </span></span>
<span><span class="co">##  [31] "modMass"                  "modLocation"             </span></span>
<span><span class="co">##  [33] "subOriginalResidue"       "subReplacementResidue"   </span></span>
<span><span class="co">##  [35] "subLocation"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Verify that this table contains 5802 matches for 5343 scans and 4938 peptides sequences.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="co">## number of matches</span></span>
<span><span class="co">##  [1] 5802</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span><span class="op">)</span> <span class="co">## number of scans</span></span>
<span><span class="co">##  [1] 5343</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span>   <span class="co">## number of peptide sequences</span></span>
<span><span class="co">##  [1] 4938</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The PSM data are read as is, without any filtering. As we can see below, we still have all the hits from the forward and reverse (decoy) databases.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">isDecoy</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  FALSE  TRUE </span></span>
<span><span class="co">##   2906  2896</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="keeping-all-matches" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="keeping-all-matches">
<span class="header-section-number">3.2</span> Keeping all matches</h2>
<p>The data contains also contains multiple matches for several spectra. The table below shows the number of number of spectra that have 1, 2, … up to 5 matches.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##     1    2    3    4    5 </span></span>
<span><span class="co">##  4936  369   26   10    2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Below, we can see how scan 1774 has 4 matches, all to sequence <code>RTRYQAEVR</code>, which itself matches to 4 different proteins:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">spectrumID</span> <span class="op">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=1774"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">id</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">##     sequence                                    spectrumID chargeState rank</span></span>
<span><span class="co">##  1 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2    1</span></span>
<span><span class="co">##  2 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2    1</span></span>
<span><span class="co">##  3 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2    1</span></span>
<span><span class="co">##  4 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2    1</span></span>
<span><span class="co">##    passThreshold</span></span>
<span><span class="co">##  1          TRUE</span></span>
<span><span class="co">##  2          TRUE</span></span>
<span><span class="co">##  3          TRUE</span></span>
<span><span class="co">##  4          TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the goal is to keep all the matches, but arranged by scan/spectrum, one can <em>reduce</em> the <code>PSM</code> object by the <code>spectrumID</code> variable, so that each scan correponds to a single row that still stores all values<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/PSM.html">reducePSMs</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">id</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span></span>
<span><span class="va">id2</span></span>
<span><span class="co">##  Reduced PSM with 5343 rows and 35 columns.</span></span>
<span><span class="co">##  names(35): sequence spectrumID ... subReplacementResidue subLocation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting object contains a single entry for scan 1774 with information for the multiple matches stored as lists within the cells.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">j</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">id2</span><span class="op">$</span><span class="va">spectrumID</span> <span class="op">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=1774"</span><span class="op">)</span></span>
<span><span class="va">id2</span><span class="op">[</span><span class="va">j</span>, <span class="op">]</span></span>
<span><span class="co">##  Reduced PSM with 1 rows and 35 columns.</span></span>
<span><span class="co">##  names(35): sequence spectrumID ... subReplacementResidue subLocation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id2</span><span class="op">[</span><span class="va">j</span>, <span class="st">"DatabaseAccess"</span><span class="op">]</span></span>
<span><span class="co">##  CharacterList of length 1</span></span>
<span><span class="co">##  [["controllerType=0 controllerNumber=1 scan=1774"]] ECA2104 ... ECA4142</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The is the type of complete identification table that could be used to annotate an raw mass spectrometry <code>Spectra</code> object, as shown below.</p>
</section><section id="filtering-data" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="filtering-data">
<span class="header-section-number">3.3</span> Filtering data</h2>
<p>Often, the PSM data is filtered to only retain reliable matches. The <code>MSnID</code> package can be used to set thresholds to attain user-defined PSM, peptide or protein-level FDRs. Here, we will simply filter out wrong identification manually.</p>
<p>Here, the <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> from the <code>dplyr</code> package comes very handy. We will thus start by converting the <code>DataFrame</code> to a <code>tibble</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">id_tbl</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">id_tbl</span></span>
<span><span class="co">##  # A tibble: 5,802 × 35</span></span>
<span><span class="co">##    sequence                  spectrumID        chargeState  rank passThreshold</span></span>
<span><span class="co">##    &lt;chr&gt;                     &lt;chr&gt;                   &lt;int&gt; &lt;int&gt; &lt;lgl&gt;        </span></span>
<span><span class="co">##  1 RQCRTDFLNYLR              controllerType=0…           3     1 TRUE         </span></span>
<span><span class="co">##  2 ESVALADQVTCVDWRNRKATKK    controllerType=0…           2     1 TRUE         </span></span>
<span><span class="co">##  3 KELLCLAMQIIR              controllerType=0…           2     1 TRUE         </span></span>
<span><span class="co">##  4 QRMARTSDKQQSIRFLERLCGR    controllerType=0…           3     1 TRUE         </span></span>
<span><span class="co">##  5 KDEGSTEPLKVVRDMTDAICMLLR  controllerType=0…           3     1 TRUE         </span></span>
<span><span class="co">##  6 DGGPAIYGHERVGRNAKKFKCLKFR controllerType=0…           3     1 TRUE         </span></span>
<span><span class="co">##  # ℹ 5,796 more rows</span></span>
<span><span class="co">##  # ℹ 30 more variables: experimentalMassToCharge &lt;dbl&gt;, …</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Remove decoy hits</li>
</ul>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_tbl</span> <span class="op">&lt;-</span> <span class="va">id_tbl</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">isDecoy</span><span class="op">)</span></span>
<span><span class="va">id_tbl</span></span>
<span><span class="co">##  # A tibble: 2,906 × 35</span></span>
<span><span class="co">##    sequence                    spectrumID      chargeState  rank passThreshold</span></span>
<span><span class="co">##    &lt;chr&gt;                       &lt;chr&gt;                 &lt;int&gt; &lt;int&gt; &lt;lgl&gt;        </span></span>
<span><span class="co">##  1 RQCRTDFLNYLR                controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  2 ESVALADQVTCVDWRNRKATKK      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  3 QRMARTSDKQQSIRFLERLCGR      controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  4 DGGPAIYGHERVGRNAKKFKCLKFR   controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  5 QRMARTSDKQQSIRFLERLCGR      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  6 CIDRARHVEVQIFGDGKGRVVALGER… controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  # ℹ 2,900 more rows</span></span>
<span><span class="co">##  # ℹ 30 more variables: experimentalMassToCharge &lt;dbl&gt;, …</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<ul>
<li>Keep first rank matches</li>
</ul>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_tbl</span> <span class="op">&lt;-</span> <span class="va">id_tbl</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">id_tbl</span></span>
<span><span class="co">##  # A tibble: 2,751 × 35</span></span>
<span><span class="co">##    sequence                    spectrumID      chargeState  rank passThreshold</span></span>
<span><span class="co">##    &lt;chr&gt;                       &lt;chr&gt;                 &lt;int&gt; &lt;int&gt; &lt;lgl&gt;        </span></span>
<span><span class="co">##  1 RQCRTDFLNYLR                controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  2 ESVALADQVTCVDWRNRKATKK      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  3 QRMARTSDKQQSIRFLERLCGR      controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  4 DGGPAIYGHERVGRNAKKFKCLKFR   controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  5 QRMARTSDKQQSIRFLERLCGR      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  6 CIDRARHVEVQIFGDGKGRVVALGER… controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  # ℹ 2,745 more rows</span></span>
<span><span class="co">##  # ℹ 30 more variables: experimentalMassToCharge &lt;dbl&gt;, …</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<ul>
<li>Remove shared peptides. Start by identifying scans that match different proteins. For example scan 4884 matches proteins <code>XXX_ECA3406</code> and <code>ECA3415</code>. Scan 4099 match <code>XXX_ECA4416_1</code>, <code>XXX_ECA4416_2</code> and <code>XXX_ECA4416_3</code>. Then remove the scans that match any of these proteins.</li>
</ul>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mltm</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">id_tbl</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">spectrumID</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>nProts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">DatabaseAccess</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">nProts</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">spectrumID</span>, <span class="va">nProts</span><span class="op">)</span></span>
<span><span class="va">mltm</span></span>
<span><span class="co">##  # A tibble: 85 × 2</span></span>
<span><span class="co">##  # Groups:   spectrumID [39]</span></span>
<span><span class="co">##    spectrumID                                    nProts</span></span>
<span><span class="co">##    &lt;chr&gt;                                          &lt;int&gt;</span></span>
<span><span class="co">##  1 controllerType=0 controllerNumber=1 scan=1073      2</span></span>
<span><span class="co">##  2 controllerType=0 controllerNumber=1 scan=1073      2</span></span>
<span><span class="co">##  3 controllerType=0 controllerNumber=1 scan=6578      2</span></span>
<span><span class="co">##  4 controllerType=0 controllerNumber=1 scan=6578      2</span></span>
<span><span class="co">##  5 controllerType=0 controllerNumber=1 scan=5617      2</span></span>
<span><span class="co">##  6 controllerType=0 controllerNumber=1 scan=5617      2</span></span>
<span><span class="co">##  # ℹ 79 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_tbl</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">id_tbl</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">spectrumID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">mltm</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span></span>
<span><span class="va">id_tbl</span></span>
<span><span class="co">##  # A tibble: 2,666 × 35</span></span>
<span><span class="co">##    sequence                    spectrumID      chargeState  rank passThreshold</span></span>
<span><span class="co">##    &lt;chr&gt;                       &lt;chr&gt;                 &lt;int&gt; &lt;int&gt; &lt;lgl&gt;        </span></span>
<span><span class="co">##  1 RQCRTDFLNYLR                controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  2 ESVALADQVTCVDWRNRKATKK      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  3 QRMARTSDKQQSIRFLERLCGR      controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  4 DGGPAIYGHERVGRNAKKFKCLKFR   controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  5 QRMARTSDKQQSIRFLERLCGR      controllerType…           2     1 TRUE         </span></span>
<span><span class="co">##  6 CIDRARHVEVQIFGDGKGRVVALGER… controllerType…           3     1 TRUE         </span></span>
<span><span class="co">##  # ℹ 2,660 more rows</span></span>
<span><span class="co">##  # ℹ 30 more variables: experimentalMassToCharge &lt;dbl&gt;, …</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Which leaves us with 2666 PSMs.</p>
<p>This can also be achieved with the <code><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html">filterPSMs()</a></code> function, or the individual <code><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html">filterPsmRank()</a></code>, <code>filterPsmDecoy</code> and <code><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html">filterPsmShared()</a></code> functions:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html">filterPSMs</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">##  Starting with 5802 PSMs:</span></span>
<span><span class="co">##  Removed 2896 decoy hits.</span></span>
<span><span class="co">##  Removed 155 PSMs with rank &gt; 1.</span></span>
<span><span class="co">##  Removed 85 shared peptides.</span></span>
<span><span class="co">##  2666 PSMs left.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describePeptides()</a></code> and <code><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describeProteins()</a></code> functions from the <code>PSMatch</code> package provide useful summaries of preptides and proteins in a PSM search result.</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describePeptides()</a></code> gives the number of unique and shared peptides and for the latter, the size of their protein groups:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describePeptides</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">)</span></span>
<span><span class="co">##  2324 peptides composed of</span></span>
<span><span class="co">##   unique peptides: 2324</span></span>
<span><span class="co">##   shared peptides (among protein):</span></span>
<span><span class="co">##    ()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describeProteins()</a></code> gives the number of proteins defined by only unique, only shared, or a mixture of unique/shared peptides:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/describeProteins.html">describeProteins</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">)</span></span>
<span><span class="co">##  1466 proteins composed of</span></span>
<span><span class="co">##   only unique peptides: 1466</span></span>
<span><span class="co">##   only shared peptides: 0</span></span>
<span><span class="co">##   unique and shared peptides: 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <a href="https://rformassspectrometry.github.io/PSMatch/articles/AdjacencyMatrix.html">Understanding protein groups with adjacency matrices</a> <code>PSMatch</code> vignette provides additional tools to explore how proteins were inferred from peptides.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compare the distribution of raw identification scores of the decoy and non-decoy hits. Interpret the figure.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MS.GF.RawScore</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">isDecoy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <em><a href="https://CRAN.R-project.org/package=tidyverse">tidyverse</a></em> tools are fit for data wrangling with identification data. Using the above identification dataframe, calculate the length of each peptide (you can use <code>nchar</code> with the peptide sequence <code>sequence</code>) and the number of peptides for each protein (defined as <code>DatabaseDescription</code>). Plot the length of the proteins against their respective number of peptides.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iddf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>peplen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span><span class="op">(</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">npeps</span> <span class="op">&lt;-</span> <span class="va">iddf</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">DatabaseAccess</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">iddf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">full_join</a></span><span class="op">(</span><span class="va">iddf</span>, <span class="va">npeps</span><span class="op">)</span></span>
<span><span class="co">##  Joining with `by = join_by(DatabaseAccess)`</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">iddf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n</span>, y <span class="op">=</span> <span class="va">DBseqLength</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/answid1-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Identifcation data wrangling.</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>If you would like to learn more about how the mzid data are handled by <code>PSMatch</code> via the <em><a href="https://bioconductor.org/packages/3.19/mzR">mzR</a></em> and <em><a href="https://bioconductor.org/packages/3.19/mzID">mzID</a></em> packages, check out the @ref(sec-id2) section in the annex.</p>
</section><section id="adding-identification-data-to-raw-data" class="level2" data-number="3.4"><h2 data-number="3.4" class="anchored" data-anchor-id="adding-identification-data-to-raw-data">
<span class="header-section-number">3.4</span> Adding identification data to raw data</h2>
<p>We are goind to use the <code>sp</code> object created in the previous chapter and the <code>id_filtered</code> variable generated above.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Generating the <code>sp</code> object from scratch 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lgatto/rpx">rpx</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra">Spectra</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: BiocParallel</span></span>
<span><span class="co">##  Loading required package: ProtGenerics</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'ProtGenerics'</span></span>
<span><span class="co">##  The following object is masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      smooth</span></span>
<span><span class="va">fn</span> <span class="op">&lt;-</span> <span class="st">"TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML"</span></span>
<span><span class="va">px</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">PXDataset</a></span><span class="op">(</span><span class="st">"PXD000001"</span><span class="op">)</span></span>
<span><span class="co">##  Loading PXD000001 from cache.</span></span>
<span><span class="va">mzf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxget</a></span><span class="op">(</span><span class="va">px</span>, <span class="va">fn</span><span class="op">)</span></span>
<span><span class="co">##  Loading TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML from cache.</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html">Spectra</a></span><span class="op">(</span><span class="va">mzf</span><span class="op">)</span></span>
<span><span class="va">sp</span></span>
<span><span class="co">##  MSn data (Spectra) with 7534 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##         msLevel     rtime scanIndex</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  1            1    0.4584         1</span></span>
<span><span class="co">##  2            1    0.9725         2</span></span>
<span><span class="co">##  3            1    1.8524         3</span></span>
<span><span class="co">##  4            1    2.7424         4</span></span>
<span><span class="co">##  5            1    3.6124         5</span></span>
<span><span class="co">##  ...        ...       ...       ...</span></span>
<span><span class="co">##  7530         2   3600.47      7530</span></span>
<span><span class="co">##  7531         2   3600.83      7531</span></span>
<span><span class="co">##  7532         2   3601.18      7532</span></span>
<span><span class="co">##  7533         2   3601.57      7533</span></span>
<span><span class="co">##  7534         2   3601.98      7534</span></span>
<span><span class="co">##   ... 33 more variables/columns.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  file(s):</span></span>
<span><span class="co">##  d177d1ab84_TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>Identification data (as a <code>DataFrame</code>) can be merged into raw data (as a <code>Spectra</code> object) by adding new spectra variables to the appropriate MS2 spectra. Scans and peptide-spectrum matches can be matched by their spectrum identifers.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Identify the spectum identifier columns in the <code>sp</code> the <code>id_filtered</code> variables.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In the raw data, it is encoded as <code>spectrumId</code>, while in the identification data, we have <code>spectrumID</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraVariables</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span>
<span><span class="co">##   [1] "msLevel"                  "rtime"                   </span></span>
<span><span class="co">##   [3] "acquisitionNum"           "scanIndex"               </span></span>
<span><span class="co">##   [5] "dataStorage"              "dataOrigin"              </span></span>
<span><span class="co">##   [7] "centroided"               "smoothed"                </span></span>
<span><span class="co">##   [9] "polarity"                 "precScanNum"             </span></span>
<span><span class="co">##  [11] "precursorMz"              "precursorIntensity"      </span></span>
<span><span class="co">##  [13] "precursorCharge"          "collisionEnergy"         </span></span>
<span><span class="co">##  [15] "isolationWindowLowerMz"   "isolationWindowTargetMz" </span></span>
<span><span class="co">##  [17] "isolationWindowUpperMz"   "peaksCount"              </span></span>
<span><span class="co">##  [19] "totIonCurrent"            "basePeakMZ"              </span></span>
<span><span class="co">##  [21] "basePeakIntensity"        "ionisationEnergy"        </span></span>
<span><span class="co">##  [23] "lowMZ"                    "highMZ"                  </span></span>
<span><span class="co">##  [25] "mergedScan"               "mergedResultScanNum"     </span></span>
<span><span class="co">##  [27] "mergedResultStartScanNum" "mergedResultEndScanNum"  </span></span>
<span><span class="co">##  [29] "injectionTime"            "filterString"            </span></span>
<span><span class="co">##  [31] "spectrumId"               "ionMobilityDriftTime"    </span></span>
<span><span class="co">##  [33] "scanWindowLowerLimit"     "scanWindowUpperLimit"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">)</span></span>
<span><span class="co">##   [1] "sequence"                 "spectrumID"              </span></span>
<span><span class="co">##   [3] "chargeState"              "rank"                    </span></span>
<span><span class="co">##   [5] "passThreshold"            "experimentalMassToCharge"</span></span>
<span><span class="co">##   [7] "calculatedMassToCharge"   "peptideRef"              </span></span>
<span><span class="co">##   [9] "modNum"                   "isDecoy"                 </span></span>
<span><span class="co">##  [11] "post"                     "pre"                     </span></span>
<span><span class="co">##  [13] "start"                    "end"                     </span></span>
<span><span class="co">##  [15] "DatabaseAccess"           "DBseqLength"             </span></span>
<span><span class="co">##  [17] "DatabaseSeq"              "DatabaseDescription"     </span></span>
<span><span class="co">##  [19] "scan.number.s."           "acquisitionNum"          </span></span>
<span><span class="co">##  [21] "spectrumFile"             "idFile"                  </span></span>
<span><span class="co">##  [23] "MS.GF.RawScore"           "MS.GF.DeNovoScore"       </span></span>
<span><span class="co">##  [25] "MS.GF.SpecEValue"         "MS.GF.EValue"            </span></span>
<span><span class="co">##  [27] "MS.GF.QValue"             "MS.GF.PepQValue"         </span></span>
<span><span class="co">##  [29] "modPeptideRef"            "modName"                 </span></span>
<span><span class="co">##  [31] "modMass"                  "modLocation"             </span></span>
<span><span class="co">##  [33] "subOriginalResidue"       "subReplacementResidue"   </span></span>
<span><span class="co">##  [35] "subLocation"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>We still have several PTMs that are matched to a single spectrum identifier:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##     1    2    3    4 </span></span>
<span><span class="co">##  2630   13    2    1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at <code>"controllerType=0 controllerNumber=1 scan=5490"</code>, the has 4 matching PSMs in detail.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">##  controllerType=0 controllerNumber=1 scan=5490 </span></span>
<span><span class="co">##                                           1903</span></span>
<span><span class="va">id_4</span> <span class="op">&lt;-</span> <span class="va">id_filtered</span><span class="op">[</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumID</span> <span class="op">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=5490"</span>, <span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">id_4</span></span>
<span><span class="co">##            sequence                                    spectrumID chargeState</span></span>
<span><span class="co">##  1 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490           3</span></span>
<span><span class="co">##  2 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490           3</span></span>
<span><span class="co">##  3 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490           3</span></span>
<span><span class="co">##  4 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490           3</span></span>
<span><span class="co">##    rank passThreshold experimentalMassToCharge calculatedMassToCharge</span></span>
<span><span class="co">##  1    1          TRUE                 698.6633               698.3315</span></span>
<span><span class="co">##  2    1          TRUE                 698.6633               698.3315</span></span>
<span><span class="co">##  3    1          TRUE                 698.6633               698.3315</span></span>
<span><span class="co">##  4    1          TRUE                 698.6633               698.3315</span></span>
<span><span class="co">##    peptideRef modNum isDecoy post pre start end DatabaseAccess DBseqLength</span></span>
<span><span class="co">##  1     Pep453      4   FALSE    C   K   127 142        ECA0668         302</span></span>
<span><span class="co">##  2     Pep453      4   FALSE    C   K   127 142        ECA0668         302</span></span>
<span><span class="co">##  3     Pep453      4   FALSE    C   K   127 142        ECA0668         302</span></span>
<span><span class="co">##  4     Pep453      4   FALSE    C   K   127 142        ECA0668         302</span></span>
<span><span class="co">##    DatabaseSeq          DatabaseDescription scan.number.s. acquisitionNum</span></span>
<span><span class="co">##  1             ECA0668 hypothetical protein           5490           5490</span></span>
<span><span class="co">##  2             ECA0668 hypothetical protein           5490           5490</span></span>
<span><span class="co">##  3             ECA0668 hypothetical protein           5490           5490</span></span>
<span><span class="co">##  4             ECA0668 hypothetical protein           5490           5490</span></span>
<span><span class="co">##                                                           spectrumFile</span></span>
<span><span class="co">##  1 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  2 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  3 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  4 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##                                                                 idFile</span></span>
<span><span class="co">##  1 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  2 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  3 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  4 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##    MS.GF.RawScore MS.GF.DeNovoScore MS.GF.SpecEValue MS.GF.EValue</span></span>
<span><span class="co">##  1            -22                79     4.555588e-07     1.307689</span></span>
<span><span class="co">##  2            -22                79     4.555588e-07     1.307689</span></span>
<span><span class="co">##  3            -22                79     4.555588e-07     1.307689</span></span>
<span><span class="co">##  4            -22                79     4.555588e-07     1.307689</span></span>
<span><span class="co">##    MS.GF.QValue MS.GF.PepQValue modPeptideRef         modName  modMass</span></span>
<span><span class="co">##  1    0.9006211       0.8901099        Pep453 Carbamidomethyl 57.02146</span></span>
<span><span class="co">##  2    0.9006211       0.8901099        Pep453 Carbamidomethyl 57.02146</span></span>
<span><span class="co">##  3    0.9006211       0.8901099        Pep453 Carbamidomethyl 57.02146</span></span>
<span><span class="co">##  4    0.9006211       0.8901099        Pep453 Carbamidomethyl 57.02146</span></span>
<span><span class="co">##    modLocation subOriginalResidue subReplacementResidue subLocation</span></span>
<span><span class="co">##  1           2               &lt;NA&gt;                  &lt;NA&gt;          NA</span></span>
<span><span class="co">##  2           5               &lt;NA&gt;                  &lt;NA&gt;          NA</span></span>
<span><span class="co">##  3          10               &lt;NA&gt;                  &lt;NA&gt;          NA</span></span>
<span><span class="co">##  4          15               &lt;NA&gt;                  &lt;NA&gt;          NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that these 4 PSMs differ by the location of the Carbamidomethyl modification.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_4</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"modName"</span>, <span class="st">"modLocation"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##            modName modLocation</span></span>
<span><span class="co">##  1 Carbamidomethyl           2</span></span>
<span><span class="co">##  2 Carbamidomethyl           5</span></span>
<span><span class="co">##  3 Carbamidomethyl          10</span></span>
<span><span class="co">##  4 Carbamidomethyl          15</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s reduce that PSM table before joining it to the <code>Spectra</code> object, to make sure we have unique one-to-one matches between the raw spectra and the PSMs.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/PSM.html">reducePSMs</a></span><span class="op">(</span><span class="va">id_filtered</span>, <span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span></span>
<span><span class="va">id_filtered</span></span>
<span><span class="co">##  Reduced PSM with 2646 rows and 35 columns.</span></span>
<span><span class="co">##  names(35): sequence spectrumID ... subReplacementResidue subLocation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These two data can thus simply be joined using:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html">joinSpectraData</a></span><span class="op">(</span><span class="va">sp</span>, <span class="va">id_filtered</span>,</span>
<span>                      by.x <span class="op">=</span> <span class="st">"spectrumId"</span>,</span>
<span>                      by.y <span class="op">=</span> <span class="st">"spectrumID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraVariables</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span>
<span><span class="co">##   [1] "msLevel"                  "rtime"                   </span></span>
<span><span class="co">##   [3] "acquisitionNum"           "scanIndex"               </span></span>
<span><span class="co">##   [5] "dataStorage"              "dataOrigin"              </span></span>
<span><span class="co">##   [7] "centroided"               "smoothed"                </span></span>
<span><span class="co">##   [9] "polarity"                 "precScanNum"             </span></span>
<span><span class="co">##  [11] "precursorMz"              "precursorIntensity"      </span></span>
<span><span class="co">##  [13] "precursorCharge"          "collisionEnergy"         </span></span>
<span><span class="co">##  [15] "isolationWindowLowerMz"   "isolationWindowTargetMz" </span></span>
<span><span class="co">##  [17] "isolationWindowUpperMz"   "peaksCount"              </span></span>
<span><span class="co">##  [19] "totIonCurrent"            "basePeakMZ"              </span></span>
<span><span class="co">##  [21] "basePeakIntensity"        "ionisationEnergy"        </span></span>
<span><span class="co">##  [23] "lowMZ"                    "highMZ"                  </span></span>
<span><span class="co">##  [25] "mergedScan"               "mergedResultScanNum"     </span></span>
<span><span class="co">##  [27] "mergedResultStartScanNum" "mergedResultEndScanNum"  </span></span>
<span><span class="co">##  [29] "injectionTime"            "filterString"            </span></span>
<span><span class="co">##  [31] "spectrumId"               "ionMobilityDriftTime"    </span></span>
<span><span class="co">##  [33] "scanWindowLowerLimit"     "scanWindowUpperLimit"    </span></span>
<span><span class="co">##  [35] "sequence"                 "chargeState"             </span></span>
<span><span class="co">##  [37] "rank"                     "passThreshold"           </span></span>
<span><span class="co">##  [39] "experimentalMassToCharge" "calculatedMassToCharge"  </span></span>
<span><span class="co">##  [41] "peptideRef"               "modNum"                  </span></span>
<span><span class="co">##  [43] "isDecoy"                  "post"                    </span></span>
<span><span class="co">##  [45] "pre"                      "start"                   </span></span>
<span><span class="co">##  [47] "end"                      "DatabaseAccess"          </span></span>
<span><span class="co">##  [49] "DBseqLength"              "DatabaseSeq"             </span></span>
<span><span class="co">##  [51] "DatabaseDescription"      "scan.number.s."          </span></span>
<span><span class="co">##  [53] "acquisitionNum.y"         "spectrumFile"            </span></span>
<span><span class="co">##  [55] "idFile"                   "MS.GF.RawScore"          </span></span>
<span><span class="co">##  [57] "MS.GF.DeNovoScore"        "MS.GF.SpecEValue"        </span></span>
<span><span class="co">##  [59] "MS.GF.EValue"             "MS.GF.QValue"            </span></span>
<span><span class="co">##  [61] "MS.GF.PepQValue"          "modPeptideRef"           </span></span>
<span><span class="co">##  [63] "modName"                  "modMass"                 </span></span>
<span><span class="co">##  [65] "modLocation"              "subOriginalResidue"      </span></span>
<span><span class="co">##  [67] "subReplacementResidue"    "subLocation"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Verify that the identification data has been added to the correct spectra.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Let’s first verify that no identification data has been added to the MS1 scans.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="va">sp</span>, <span class="fl">1</span><span class="op">)</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>They have indeed been added to 56% of the MS2 spectra.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="va">sp</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">sp_2</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  FALSE  TRUE </span></span>
<span><span class="co">##   2646  3457</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s compare the precursor/peptide mass to charges</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp_2</span> <span class="op">&lt;-</span> <span class="va">sp_2</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">sp_2</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">sp_2</span><span class="op">$</span><span class="va">precursorMz</span> <span class="op">-</span> <span class="va">sp_2</span><span class="op">$</span><span class="va">experimentalMassToCharge</span><span class="op">)</span></span>
<span><span class="co">##     Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##   0.0000  0.0000  0.0000  0.0053  0.0000  2.0297</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="an-identification-annotated-chromatogram" class="level2" data-number="3.5"><h2 data-number="3.5" class="anchored" data-anchor-id="an-identification-annotated-chromatogram">
<span class="header-section-number">3.5</span> An identification-annotated chromatogram</h2>
<p>Now that we have combined raw data and their associated peptide-spectrum matches, we can produce an improved total ion chromatogram, identifying MS1 scans that lead to successful identifications.</p>
<p>The <code><a href="https://rdrr.io/pkg/Spectra/man/countIdentifications.html">countIdentifications()</a></code> function is going to tally the number of identifications (i.e non-missing characters in the <code>sequence</code> spectra variable) for each scan. In the case of MS2 scans, these will be either 1 or 0, depending on the presence of a sequence. For MS1 scans, the function will count the number of sequences for the descendant MS2 scans, i.e.&nbsp;those produced from precursor ions from each MS1 scan.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/countIdentifications.html">countIdentifications</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Below, we see on the second line that 3457 MS2 scans lead to no PSM, while 2546 lead to an identification. Among all MS1 scans, 833 lead to no MS2 scans with PSMs. 30 MS1 scans generated one MS2 scan that lead to a PSM, 45 lead to two PSMs, …</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">msLevel</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span>, <span class="va">sp</span><span class="op">$</span><span class="va">countIdentifications</span><span class="op">)</span></span>
<span><span class="co">##     </span></span>
<span><span class="co">##         0    1    2    3    4    5    6    7    8    9   10</span></span>
<span><span class="co">##    1  833   30   45   97  139  132   92   42   17    3    1</span></span>
<span><span class="co">##    2 3457 2646    0    0    0    0    0    0    0    0    0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These data can also be visualised on the total ion chromatogram:</p>
<div class="cell" data-layout-align="center" data-fig.fullwidth="true">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rtime</span>,</span>
<span>            y <span class="op">=</span> <span class="va">totIonCurrent</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">countIdentifications</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>                                   <span class="cn">NA</span>, <span class="va">countIdentifications</span><span class="op">)</span><span class="op">)</span>,</span>
<span>               size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>               alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"Number of ids"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/nSequencePlot-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section><section id="visualising-peptide-spectrum-matches" class="level2" data-number="3.6"><h2 data-number="3.6" class="anchored" data-anchor-id="visualising-peptide-spectrum-matches">
<span class="header-section-number">3.6</span> Visualising peptide-spectrum matches</h2>
<p>Let’s choose a MS2 spectrum with a high identification score and plot it.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">MS.GF.RawScore</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We have seen above that we can add labels to each peak using the <code>labels</code> argument in <code><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra()</a></code>. The <code><a href="https://rdrr.io/pkg/PSMatch/man/addFragments.html">addFragments()</a></code> function takes a spectrum as input (that is a <code>Spectra</code> object of length 1) and annotates its peaks.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/addFragments.html">addFragments</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##    [1] NA     NA     NA     "b1"   NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [11] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [21] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [31] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [41] NA     NA     NA     "y1_"  NA     NA     NA     NA     NA     "y1"  </span></span>
<span><span class="co">##   [51] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [61] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [71] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##   [81] NA     NA     NA     NA     NA     NA     "b2"   NA     NA     NA    </span></span>
<span><span class="co">##   [91] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [101] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [111] NA     NA     NA     NA     NA     NA     NA     "y2_"  "y2*"  NA    </span></span>
<span><span class="co">##  [121] NA     "y2"   NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [131] NA     NA     "b3"   NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [141] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [151] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [161] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [171] "y3*"  NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [181] "y3"   NA     "b4_"  NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [191] NA     NA     "b4"   NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [201] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [211] NA     NA     NA     NA     NA     NA     NA     "y4*"  NA     "b5_" </span></span>
<span><span class="co">##  [221] "b5*"  NA     "y4"   NA     "b5"   NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [231] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [241] NA     NA     NA     NA     NA     NA     "y5_"  "y5*"  NA     "b6_" </span></span>
<span><span class="co">##  [251] "y5"   NA     NA     "b6"   NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [261] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [271] NA     "y6*"  "b7_"  NA     "y6"   NA     NA     "b7"   NA     NA    </span></span>
<span><span class="co">##  [281] NA     NA     NA     NA     NA     NA     "y7_"  "y7*"  NA     "y7"  </span></span>
<span><span class="co">##  [291] NA     "b8"   NA     NA     NA     NA     "y8_"  "y8*"  "y8"   NA    </span></span>
<span><span class="co">##  [301] NA     NA     NA     NA     NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [311] "y9_"  "y9*"  NA     "y9"   NA     NA     NA     NA     NA     NA    </span></span>
<span><span class="co">##  [321] "y10_" "y10*" NA     "y10"  NA     NA     NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It can be directly used with <code><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra()</a></code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">addFragments</span>,</span>
<span>            labelPos <span class="op">=</span> <span class="fl">3</span>, labelCol <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>When a precursor peptide ion is fragmented in a CID cell, it breaks at specific bonds, producing sets of peaks (<em>a</em>, <em>b</em>, <em>c</em> and <em>x</em>, <em>y</em>, <em>z</em>) that can be predicted.</p>
<div class="cell" data-layout-align="center" data-fig.margin="false">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="img/frag.png" class="img-fluid figure-img" style="width:80.0%"></p>
<figcaption class="figure-caption">Peptide fragmentation.</figcaption></figure>
</div>
</div>
</div>
<p>The annotation of spectra is obtained by simulating fragmentation of a peptide and matching observed peaks to fragments:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">$</span><span class="va">sequence</span></span>
<span><span class="co">##  [1] "THSQEEMQHMQR"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">calculateFragments</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span></span>
<span><span class="co">##  Modifications used: C=57.02146</span></span>
<span><span class="co">##            mz  ion type pos z         seq</span></span>
<span><span class="co">##  1   102.0550   b1    b   1 1           T</span></span>
<span><span class="co">##  2   239.1139   b2    b   2 1          TH</span></span>
<span><span class="co">##  3   326.1459   b3    b   3 1         THS</span></span>
<span><span class="co">##  4   454.2045   b4    b   4 1        THSQ</span></span>
<span><span class="co">##  5   583.2471   b5    b   5 1       THSQE</span></span>
<span><span class="co">##  6   712.2897   b6    b   6 1      THSQEE</span></span>
<span><span class="co">##  7   843.3301   b7    b   7 1     THSQEEM</span></span>
<span><span class="co">##  8   971.3887   b8    b   8 1    THSQEEMQ</span></span>
<span><span class="co">##  9  1108.4476   b9    b   9 1   THSQEEMQH</span></span>
<span><span class="co">##  10 1239.4881  b10    b  10 1  THSQEEMQHM</span></span>
<span><span class="co">##  11 1367.5467  b11    b  11 1 THSQEEMQHMQ</span></span>
<span><span class="co">##  12  175.1190   y1    y   1 1           R</span></span>
<span><span class="co">##  13  303.1775   y2    y   2 1          QR</span></span>
<span><span class="co">##  14  434.2180   y3    y   3 1         MQR</span></span>
<span><span class="co">##  15  571.2769   y4    y   4 1        HMQR</span></span>
<span><span class="co">##  16  699.3355   y5    y   5 1       QHMQR</span></span>
<span><span class="co">##  17  830.3760   y6    y   6 1      MQHMQR</span></span>
<span><span class="co">##  18  959.4186   y7    y   7 1     EMQHMQR</span></span>
<span><span class="co">##  19 1088.4612   y8    y   8 1    EEMQHMQR</span></span>
<span><span class="co">##  20 1216.5198   y9    y   9 1   QEEMQHMQR</span></span>
<span><span class="co">##  21 1303.5518  y10    y  10 1  SQEEMQHMQR</span></span>
<span><span class="co">##  22 1440.6107  y11    y  11 1 HSQEEMQHMQR</span></span>
<span><span class="co">##  23  436.1939  b4_   b_   4 1        THSQ</span></span>
<span><span class="co">##  24  565.2365  b5_   b_   5 1       THSQE</span></span>
<span><span class="co">##  25  694.2791  b6_   b_   6 1      THSQEE</span></span>
<span><span class="co">##  26  825.3196  b7_   b_   7 1     THSQEEM</span></span>
<span><span class="co">##  27  953.3782  b8_   b_   8 1    THSQEEMQ</span></span>
<span><span class="co">##  28 1090.4371  b9_   b_   9 1   THSQEEMQH</span></span>
<span><span class="co">##  29 1221.4776 b10_   b_  10 1  THSQEEMQHM</span></span>
<span><span class="co">##  30 1349.5361 b11_   b_  11 1 THSQEEMQHMQ</span></span>
<span><span class="co">##  31  941.4080  y7_   y_   7 1     EMQHMQR</span></span>
<span><span class="co">##  32 1070.4506  y8_   y_   8 1    EEMQHMQR</span></span>
<span><span class="co">##  33 1422.6001 y11_   y_  11 1 HSQEEMQHMQR</span></span>
<span><span class="co">##  34  157.1084  y1_   y_   1 1           R</span></span>
<span><span class="co">##  35  285.1670  y2_   y_   2 1          QR</span></span>
<span><span class="co">##  36  416.2075  y3_   y_   3 1         MQR</span></span>
<span><span class="co">##  37  553.2664  y4_   y_   4 1        HMQR</span></span>
<span><span class="co">##  38  681.3249  y5_   y_   5 1       QHMQR</span></span>
<span><span class="co">##  39  812.3654  y6_   y_   6 1      MQHMQR</span></span>
<span><span class="co">##  40 1198.5092  y9_   y_   9 1   QEEMQHMQR</span></span>
<span><span class="co">##  41 1285.5412 y10_   y_  10 1  SQEEMQHMQR</span></span>
<span><span class="co">##  42  566.2205  b5*   b*   5 1       THSQE</span></span>
<span><span class="co">##  43  695.2631  b6*   b*   6 1      THSQEE</span></span>
<span><span class="co">##  44  826.3036  b7*   b*   7 1     THSQEEM</span></span>
<span><span class="co">##  45  954.3622  b8*   b*   8 1    THSQEEMQ</span></span>
<span><span class="co">##  46 1091.4211  b9*   b*   9 1   THSQEEMQH</span></span>
<span><span class="co">##  47 1222.4616 b10*   b*  10 1  THSQEEMQHM</span></span>
<span><span class="co">##  48 1350.5202 b11*   b*  11 1 THSQEEMQHMQ</span></span>
<span><span class="co">##  49  286.1510  y2*   y*   2 1          QR</span></span>
<span><span class="co">##  50  417.1915  y3*   y*   3 1         MQR</span></span>
<span><span class="co">##  51  554.2504  y4*   y*   4 1        HMQR</span></span>
<span><span class="co">##  52  682.3090  y5*   y*   5 1       QHMQR</span></span>
<span><span class="co">##  53  813.3495  y6*   y*   6 1      MQHMQR</span></span>
<span><span class="co">##  54  942.3920  y7*   y*   7 1     EMQHMQR</span></span>
<span><span class="co">##  55 1071.4346  y8*   y*   8 1    EEMQHMQR</span></span>
<span><span class="co">##  56 1199.4932  y9*   y*   9 1   QEEMQHMQR</span></span>
<span><span class="co">##  57 1286.5252 y10*   y*  10 1  SQEEMQHMQR</span></span>
<span><span class="co">##  58 1423.5842 y11*   y*  11 1 HSQEEMQHMQR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="comparing-spectra" class="level2" data-number="3.7"><h2 data-number="3.7" class="anchored" data-anchor-id="comparing-spectra">
<span class="header-section-number">3.7</span> Comparing spectra</h2>
<p>The <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">compareSpectra()</a></code> function can be used to compare spectra (by default, computing the normalised dot product).</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a new <code>Spectra</code> object containing the MS2 spectra with sequences <code>"SQILQQAGTSVLSQANQVPQTVLSLLR"</code> and <code>"TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"</code>.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">sequence</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SQILQQAGTSVLSQANQVPQTVLSLLR"</span>, <span class="st">"TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp_k</span> <span class="op">&lt;-</span> <span class="va">sp</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span><span class="va">sp_k</span></span>
<span><span class="co">##  MSn data (Spectra) with 5 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##      msLevel     rtime scanIndex</span></span>
<span><span class="co">##    &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  1         2   2687.42      5230</span></span>
<span><span class="co">##  2         2   2688.88      5235</span></span>
<span><span class="co">##  3         2   2748.75      5397</span></span>
<span><span class="co">##  4         2   2765.26      5442</span></span>
<span><span class="co">##  5         2   2768.17      5449</span></span>
<span><span class="co">##   ... 68 more variables/columns.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  file(s):</span></span>
<span><span class="co">##  d177d1ab84_TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Calculate the 5 by 5 similarity matrix between all spectra using <code>compareSpectra</code>. See the <code><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html">?Spectra</a></code> man page for details. Draw a heatmap of that matrix.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">compareSpectra</a></span><span class="op">(</span><span class="va">sp_k</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strtrim.html">strtrim</a></span><span class="op">(</span><span class="va">sp_k</span><span class="op">$</span><span class="va">sequence</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mat</span></span>
<span><span class="co">##               TK          TK           SQ          SQ           SQ</span></span>
<span><span class="co">##  TK 1.0000000000 0.109126094 0.0009373465 0.001261338 0.0008256185</span></span>
<span><span class="co">##  TK 0.1091260942 1.000000000 0.0025314670 0.001459654 0.0017613212</span></span>
<span><span class="co">##  SQ 0.0009373465 0.002531467 1.0000000000 0.432133016 0.6879331218</span></span>
<span><span class="co">##  SQ 0.0012613380 0.001459654 0.4321330158 1.000000000 0.4467153012</span></span>
<span><span class="co">##  SQ 0.0008256185 0.001761321 0.6879331218 0.446715301 1.0000000000</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compare the spectra with the plotting function seen previously.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterIntensity</a></span><span class="op">(</span><span class="va">sp_k</span>, <span class="fl">1e3</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra</a></span><span class="op">(</span>main <span class="op">=</span> <span class="va">sp_k</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectraMirror</a></span><span class="op">(</span><span class="va">sp_k</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">sp_k</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"TK..."</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectraMirror</a></span><span class="op">(</span><span class="va">sp_k</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">sp_k</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"SQ..."</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectraMirror</a></span><span class="op">(</span><span class="va">sp_k</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">sp_k</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"SQ..."</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-33-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="summary-exercise" class="level2" data-number="3.8"><h2 data-number="3.8" class="anchored" data-anchor-id="summary-exercise">
<span class="header-section-number">3.8</span> Summary exercise</h2>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Download the 3 first mzML and mzID files from the <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD022816">PXD022816</a> project <span class="citation" data-cites="Morgenstern:2020">(<a href="#ref-Morgenstern:2020" role="doc-biblioref">Morgenstern, Barzilay, and Levin 2021</a>)</span>.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Getting data from PX/PRIDE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lgatto/rpx">rpx</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## https://www.ebi.ac.uk/pride/archive/projects/PXD022816</span></span>
<span><span class="co">## RawBeans: A Simple, Vendor-Independent, Raw-Data Quality-Control</span></span>
<span><span class="co">## Tool ()</span></span>
<span></span>
<span><span class="va">PXD022816</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">PXDataset</a></span><span class="op">(</span><span class="st">"PXD022816"</span><span class="op">)</span></span>
<span><span class="co">##  Querying ProteomeXchange for PXD022816.</span></span>
<span><span class="va">PXD022816</span></span>
<span><span class="co">##  Project PXD022816 with 32 files</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Resource ID BFC3 in cache in /root/.cache/R/rpx.</span></span>
<span><span class="co">##   [1] 'QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz' ... [32] 'checksum.txt'</span></span>
<span><span class="co">##   Use 'pxfiles(.)' to see all files.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxfiles</a></span><span class="op">(</span><span class="va">PXD022816</span><span class="op">)</span></span>
<span><span class="co">##  Project PXD022816 files (32):</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_04-calib.mzID.gz</span></span>
<span><span class="co">##   ...</span></span>
<span></span>
<span><span class="op">(</span><span class="va">mzids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxget</a></span><span class="op">(</span><span class="va">PXD022816</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"mzID"</span>, <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxfiles</a></span><span class="op">(</span><span class="va">PXD022816</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  Project PXD022816 files (32):</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_04-calib.mzID.gz</span></span>
<span><span class="co">##   ...</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz file.</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz file.</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz file.</span></span>
<span><span class="co">##  [1] "/root/.cache/R/rpx/15c31fe8b54_QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz"</span></span>
<span><span class="co">##  [2] "/root/.cache/R/rpx/15c41f88c10_QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz"</span></span>
<span><span class="co">##  [3] "/root/.cache/R/rpx/15c6ad49c14_QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz"</span></span>
<span><span class="op">(</span><span class="va">mzmls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxget</a></span><span class="op">(</span><span class="va">PXD022816</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"mzML"</span>, <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html">pxfiles</a></span><span class="op">(</span><span class="va">PXD022816</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  Project PXD022816 files (32):</span></span>
<span><span class="co">##   [local]  QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_01.raw</span></span>
<span><span class="co">##   [local]  QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_02.raw</span></span>
<span><span class="co">##   [local]  QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_03.raw</span></span>
<span><span class="co">##   [remote] QEP2LC6_HeLa_50ng_251120_04-calib.mzID.gz</span></span>
<span><span class="co">##   ...</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_01-calib.mzML file.</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_02-calib.mzML file.</span></span>
<span><span class="co">##  Downloading QEP2LC6_HeLa_50ng_251120_03-calib.mzML file.</span></span>
<span><span class="co">##  [1] "/root/.cache/R/rpx/15c1af8434c_QEP2LC6_HeLa_50ng_251120_01-calib.mzML"</span></span>
<span><span class="co">##  [2] "/root/.cache/R/rpx/15c2c4818d9_QEP2LC6_HeLa_50ng_251120_02-calib.mzML"</span></span>
<span><span class="co">##  [3] "/root/.cache/R/rpx/15c584379d5_QEP2LC6_HeLa_50ng_251120_03-calib.mzML"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate a <code>Spectra</code> object and a table of filtered PSMs. Visualise the total ion chromatograms and check the quality of the identification data by comparing the density of the decoy and target PSMs id scores for each file.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-24-contents" aria-controls="callout-24" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-24" class="callout-24-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Loading raw data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra">Spectra</a></span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html">Spectra</a></span><span class="op">(</span><span class="va">mzmls</span><span class="op">)</span></span>
<span><span class="va">sp</span></span>
<span><span class="co">##  MSn data (Spectra) with 87647 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##          msLevel     rtime scanIndex</span></span>
<span><span class="co">##        &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  1             1  0.177987         1</span></span>
<span><span class="co">##  2             1  0.599870         2</span></span>
<span><span class="co">##  3             1  0.978849         3</span></span>
<span><span class="co">##  4             1  1.363217         4</span></span>
<span><span class="co">##  5             1  1.742965         5</span></span>
<span><span class="co">##  ...         ...       ...       ...</span></span>
<span><span class="co">##  87643         1   4198.64     28736</span></span>
<span><span class="co">##  87644         1   4199.02     28737</span></span>
<span><span class="co">##  87645         2   4199.28     28738</span></span>
<span><span class="co">##  87646         1   4199.44     28739</span></span>
<span><span class="co">##  87647         1   4199.82     28740</span></span>
<span><span class="co">##   ... 33 more variables/columns.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  file(s):</span></span>
<span><span class="co">##  15c1af8434c_QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##  15c2c4818d9_QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##  15c584379d5_QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span></span>
<span><span class="co">## number of spectra per file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">dataOrigin</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  15c1af8434c_QEP2LC6_HeLa_50ng_251120_01-calib.mzML </span></span>
<span><span class="co">##                                               29575 </span></span>
<span><span class="co">##  15c2c4818d9_QEP2LC6_HeLa_50ng_251120_02-calib.mzML </span></span>
<span><span class="co">##                                               29332 </span></span>
<span><span class="co">##  15c584379d5_QEP2LC6_HeLa_50ng_251120_03-calib.mzML </span></span>
<span><span class="co">##                                               28740</span></span>
<span></span>
<span><span class="co">## all levels are centroided</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">centroided</span>, <span class="va">sp</span><span class="op">$</span><span class="va">msLevel</span><span class="op">)</span></span>
<span><span class="co">##        </span></span>
<span><span class="co">##             1     2</span></span>
<span><span class="co">##    TRUE 19607 68040</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Chromatograms</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="va">sp</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rtime</span>,</span>
<span>               y <span class="op">=</span> <span class="va">totIonCurrent</span>,</span>
<span>               colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">dataOrigin</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Identification data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/PSM">PSMatch</a></span><span class="op">)</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/PSM.html">PSM</a></span><span class="op">(</span><span class="va">mzids</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Number of PSMs per acquisition</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">id</span><span class="op">$</span><span class="va">idFile</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  15c31fe8b54_QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz </span></span>
<span><span class="co">##                                                  25231 </span></span>
<span><span class="co">##  15c41f88c10_QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz </span></span>
<span><span class="co">##                                                  25083 </span></span>
<span><span class="co">##  15c6ad49c14_QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz </span></span>
<span><span class="co">##                                                  24436</span></span>
<span></span>
<span><span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MetaMorpheus.score</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">isDecoy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">spectrumFile</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">id_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html">filterPSMs</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">##  Starting with 74750 PSMs:</span></span>
<span><span class="co">##  Removed 543 decoy hits.</span></span>
<span><span class="co">##  Removed 0 PSMs with rank &gt; 1.</span></span>
<span><span class="co">##  Removed 11310 shared peptides.</span></span>
<span><span class="co">##  62897 PSMs left.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">PSM.level.q.value</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0.009994817</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Join the raw and identification data. Beware though that the joining must now be performed by spectrum ids and by files.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## primary key for spectra</span></span>
<span><span class="va">sp</span><span class="op">$</span><span class="va">pkey</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"^.+_QEP"</span>, <span class="st">"QEP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">dataOrigin</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"^.+="</span>, <span class="st">"::"</span>, <span class="va">sp</span><span class="op">$</span><span class="va">spectrumId</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">pkey</span><span class="op">)</span></span>
<span><span class="co">##  [1] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::1"</span></span>
<span><span class="co">##  [2] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::2"</span></span>
<span><span class="co">##  [3] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::3"</span></span>
<span><span class="co">##  [4] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::4"</span></span>
<span><span class="co">##  [5] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::5"</span></span>
<span><span class="co">##  [6] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::6"</span></span>
<span></span>
<span><span class="co">## primary key for PSMs</span></span>
<span><span class="va">id_filtered</span><span class="op">$</span><span class="va">pkey</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"^.+\\QEP"</span>, <span class="st">"QEP"</span>, <span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumFile</span><span class="op">)</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"^.+="</span>, <span class="st">"::"</span>, <span class="va">id_filtered</span><span class="op">$</span><span class="va">spectrumID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">pkey</span><span class="op">)</span></span>
<span><span class="co">##  [1] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::3426" </span></span>
<span><span class="co">##  [2] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::20165"</span></span>
<span><span class="co">##  [3] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::20180"</span></span>
<span><span class="co">##  [4] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::15180"</span></span>
<span><span class="co">##  [5] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::10327"</span></span>
<span><span class="co">##  [6] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::12894"</span></span>
<span></span>
<span><span class="co">## For simplicity, let keep single hits per spectrum id.</span></span>
<span><span class="co">## Alternatively, explore duplicates and use QFeatures::reduceDataFrame</span></span>
<span><span class="va">id_filtered</span> <span class="op">&lt;-</span> <span class="va">id_filtered</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">pkey</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">id_filtered</span><span class="op">$</span><span class="va">pkey</span><span class="op">)</span></span>
<span><span class="co">##  [1] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::3426" </span></span>
<span><span class="co">##  [2] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::20165"</span></span>
<span><span class="co">##  [3] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::20180"</span></span>
<span><span class="co">##  [4] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::15180"</span></span>
<span><span class="co">##  [5] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::10327"</span></span>
<span><span class="co">##  [6] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML::12894"</span></span>
<span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html">joinSpectraData</a></span><span class="op">(</span><span class="va">sp</span>, <span class="va">id_filtered</span>, by.x <span class="op">=</span> <span class="st">"pkey"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Number of MS2 scans with a PSM</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="va">sp</span>, <span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  FALSE  TRUE </span></span>
<span><span class="co">##  22006 46034</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Extract the PSMs that have been matched to peptides from protein <code>O43175</code> and compare and cluster the scans. Hint: once you have created the smaller <code>Spectra</code> object with the scans of interest, switch to an in-memory backend to seed up the calculations.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp_O43175</span> <span class="op">&lt;-</span> <span class="va">sp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">DatabaseAccess</span> <span class="op">==</span> <span class="st">"O43175"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sp_O43175</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/hidden_aliases.html">setBackend</a></span><span class="op">(</span><span class="va">sp_O43175</span>, <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/MsBackend.html">MsBackendDataFrame</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp_O43175</span></span>
<span><span class="co">##  MSn data (Spectra) with 77 spectra in a MsBackendDataFrame backend:</span></span>
<span><span class="co">##        msLevel     rtime scanIndex</span></span>
<span><span class="co">##      &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  1           2   1496.36      8143</span></span>
<span><span class="co">##  2           2   1518.19      8362</span></span>
<span><span class="co">##  3           2   1582.59      9009</span></span>
<span><span class="co">##  4           2   1587.08      9056</span></span>
<span><span class="co">##  5           2   1708.69     10288</span></span>
<span><span class="co">##  ...       ...       ...       ...</span></span>
<span><span class="co">##  73          2   3148.28     23503</span></span>
<span><span class="co">##  74          2   3242.61     24333</span></span>
<span><span class="co">##  75          2   3436.78     25513</span></span>
<span><span class="co">##  76          2   3547.90     26170</span></span>
<span><span class="co">##  77          2   3563.78     26314</span></span>
<span><span class="co">##   ... 65 more variables/columns.</span></span>
<span><span class="co">##  Processing:</span></span>
<span><span class="co">##   Switch backend from MsBackendMzR to MsBackendDataFrame [Tue Dec  5 16:31:49 2023]</span></span>
<span></span>
<span><span class="va">cmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">compareSpectra</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cmat</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cmat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strtrim.html">strtrim</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">$</span><span class="va">sequence</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">cmat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="op">(</span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cmat</span><span class="op">)</span> <span class="op">==</span> <span class="st">"DLP"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 21 23 45 46 49 73 74 77</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectra</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">addFragments</span>,</span>
<span>            labelPos <span class="op">=</span> <span class="fl">3</span>, labelCol <span class="op">=</span> <span class="st">"steelblue"</span>,</span>
<span>            main <span class="op">=</span> <span class="va">sp_O43175</span><span class="op">$</span><span class="va">sequence</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">precursorCharge</span></span>
<span><span class="co">##  [1] 2 2 2 2 2 2 2 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">precursorMz</span></span>
<span><span class="co">##  [1] 493.8055 515.3086 493.8383 493.8380 515.3084 493.8057 493.8390 515.3087</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">modName</span></span>
<span><span class="co">##  [1] NA         "Carbamyl" NA         NA         "Carbamyl" NA        </span></span>
<span><span class="co">##  [7] NA         "Carbamyl"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-40-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Directly compare spectra with/without Carbamyl</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectraMirror</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="va">sp_O43175</span><span class="op">[</span><span class="fl">9</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html">plotSpectraMirror</a></span><span class="op">(</span><span class="va">sp_O43175</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">sp_O43175</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-41-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate total ion chromatograms for each acquisition and annotate the MS1 scans with the number of PSMs using the <code><a href="https://rdrr.io/pkg/Spectra/man/countIdentifications.html">countIdentifications()</a></code> function, as shown above. The function will automatically perform the counts in parallel for each acquisition.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-30-contents" aria-controls="callout-30" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-30" class="callout-30-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/countIdentifications.html">countIdentifications</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">msLevel</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span>, <span class="va">sp</span><span class="op">$</span><span class="va">countIdentifications</span><span class="op">)</span></span>
<span><span class="co">##     </span></span>
<span><span class="co">##          0     1     2     3     4     5     6     7     8     9    10</span></span>
<span><span class="co">##    1 12937   488   254   213   237   345   588   991  1439  1401   714</span></span>
<span><span class="co">##    2 22006 46034     0     0     0     0     0     0     0     0     0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterMsLevel</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">spectraData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rtime</span>, y <span class="op">=</span> <span class="va">totIonCurrent</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">countIdentifications</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>                                <span class="cn">NA</span>, <span class="va">countIdentifications</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>            alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"orange"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"^.+_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">dataOrigin</span><span class="op">)</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"Number of ids"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="20-id_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="exploration-and-assessment-of-identifications-using-msnid" class="level2" data-number="3.9"><h2 data-number="3.9" class="anchored" data-anchor-id="exploration-and-assessment-of-identifications-using-msnid">
<span class="header-section-number">3.9</span> Exploration and Assessment of Identifications using <code>MSnID</code>
</h2>
<p>The <code>MSnID</code> package extracts MS/MS ID data from mzIdentML (leveraging the <code>mzID</code> package) or text files. After collating the search results from multiple datasets it assesses their identification quality and optimises filtering criteria to achieve the maximum number of identifications while not exceeding a specified false discovery rate. It also contains a number of utilities to explore the MS/MS results and assess missed and irregular enzymatic cleavages, mass measurement accuracy, etc.</p>
<section id="step-by-step-work-flow" class="level3" data-number="3.9.1"><h3 data-number="3.9.1" class="anchored" data-anchor-id="step-by-step-work-flow">
<span class="header-section-number">3.9.1</span> Step-by-step work-flow</h3>
<p>Let’s reproduce parts of the analysis described the <code>MSnID</code> vignette. You can explore more with</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html">vignette</a></span><span class="op">(</span><span class="st">"msnid_vignette"</span>, package <span class="op">=</span> <span class="st">"MSnID"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <em><a href="https://bioconductor.org/packages/3.19/MSnID">MSnID</a></em> package can be used for post-search filtering of MS/MS identifications. One starts with the construction of an <code>MSnID</code> object that is populated with identification results that can be imported from a <code>data.frame</code> or from <code>mzIdenML</code> files. Here, we will use the example identification data provided with the package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mzids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"c_elegans.mzid.gz"</span>, package<span class="op">=</span><span class="st">"MSnID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">mzids</span><span class="op">)</span></span>
<span><span class="co">##  [1] "c_elegans.mzid.gz"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We start by loading the package, initialising the <code>MSnID</code> object, and add the identification result from our <code>mzid</code> file (there could of course be more than one).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MSnID</span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: Rcpp</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'MSnID'</span></span>
<span><span class="co">##  The following object is masked from 'package:ProtGenerics':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      peptides</span></span>
<span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/MSnID-class.html">MSnID</a></span><span class="op">(</span><span class="st">"."</span><span class="op">)</span></span>
<span><span class="co">##  Note, the anticipated/suggested columns in the</span></span>
<span><span class="co">##  peptide-to-spectrum matching results are:</span></span>
<span><span class="co">##  -----------------------------------------------</span></span>
<span><span class="co">##  accession</span></span>
<span><span class="co">##  calculatedMassToCharge</span></span>
<span><span class="co">##  chargeState</span></span>
<span><span class="co">##  experimentalMassToCharge</span></span>
<span><span class="co">##  isDecoy</span></span>
<span><span class="co">##  peptide</span></span>
<span><span class="co">##  spectrumFile</span></span>
<span><span class="co">##  spectrumID</span></span>
<span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/read_mzIDs.html">read_mzIDs</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="va">mzids</span><span class="op">)</span></span>
<span><span class="co">##  Reading from mzIdentMLs ...</span></span>
<span><span class="co">##  reading c_elegans.mzid.gz... DONE!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="co">##  MSnID object</span></span>
<span><span class="co">##  Working directory: "."</span></span>
<span><span class="co">##  #Spectrum Files:  1 </span></span>
<span><span class="co">##  #PSMs: 12263 at 36 % FDR</span></span>
<span><span class="co">##  #peptides: 9489 at 44 % FDR</span></span>
<span><span class="co">##  #accessions: 7414 at 76 % FDR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Printing the <code>MSnID</code> object returns some basic information such as</p>
<ul>
<li>Working directory.</li>
<li>Number of spectrum files used to generate data.</li>
<li>Number of peptide-to-spectrum matches and corresponding FDR.</li>
<li>Number of unique peptide sequences and corresponding FDR.</li>
<li>Number of unique proteins or amino acid sequence accessions and corresponding FDR.</li>
</ul>
<p>The package then enables to define, optimise and apply filtering based for example on missed cleavages, identification scores, precursor mass errors, etc. and assess PSM, peptide and protein FDR levels. To properly function, it expects to have access to the following data</p>
<div class="cell" data-layout-align="center">
<pre><code>##  [1] "accession"                "calculatedMassToCharge"  
##  [3] "chargeState"              "experimentalMassToCharge"
##  [5] "isDecoy"                  "peptide"                 
##  [7] "spectrumFile"             "spectrumID"</code></pre>
</div>
<p>which are indeed present in our data:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="co">##   [1] "spectrumID"                "scan number(s)"           </span></span>
<span><span class="co">##   [3] "acquisitionNum"            "passThreshold"            </span></span>
<span><span class="co">##   [5] "rank"                      "calculatedMassToCharge"   </span></span>
<span><span class="co">##   [7] "experimentalMassToCharge"  "chargeState"              </span></span>
<span><span class="co">##   [9] "MS-GF:DeNovoScore"         "MS-GF:EValue"             </span></span>
<span><span class="co">##  [11] "MS-GF:PepQValue"           "MS-GF:QValue"             </span></span>
<span><span class="co">##  [13] "MS-GF:RawScore"            "MS-GF:SpecEValue"         </span></span>
<span><span class="co">##  [15] "AssumedDissociationMethod" "IsotopeError"             </span></span>
<span><span class="co">##  [17] "isDecoy"                   "post"                     </span></span>
<span><span class="co">##  [19] "pre"                       "end"                      </span></span>
<span><span class="co">##  [21] "start"                     "accession"                </span></span>
<span><span class="co">##  [23] "length"                    "description"              </span></span>
<span><span class="co">##  [25] "pepSeq"                    "modified"                 </span></span>
<span><span class="co">##  [27] "modification"              "idFile"                   </span></span>
<span><span class="co">##  [29] "spectrumFile"              "databaseFile"             </span></span>
<span><span class="co">##  [31] "peptide"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we summarise a few steps and redirect the reader to the package’s vignette for more details:</p>
</section><section id="analysis-of-peptide-sequences" class="level3" data-number="3.9.2"><h3 data-number="3.9.2" class="anchored" data-anchor-id="analysis-of-peptide-sequences">
<span class="header-section-number">3.9.2</span> Analysis of peptide sequences</h3>
<p>Cleaning irregular cleavages at the termini of the peptides and missing cleavage site within the peptide sequences. The following two function calls create the new <code>numMisCleavages</code> and <code>numIrregCleavages</code> columns in the <code>MSnID</code> object</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/assess_termini.html">assess_termini</a></span><span class="op">(</span><span class="va">msnid</span>, validCleavagePattern<span class="op">=</span><span class="st">"[KR]\\.[^P]"</span><span class="op">)</span></span>
<span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/assess_missed_cleavages.html">assess_missed_cleavages</a></span><span class="op">(</span><span class="va">msnid</span>, missedCleavagePattern<span class="op">=</span><span class="st">"[KR](?=[^P$])"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="trimming-the-data" class="level3" data-number="3.9.3"><h3 data-number="3.9.3" class="anchored" data-anchor-id="trimming-the-data">
<span class="header-section-number">3.9.3</span> Trimming the data</h3>
<p>Now, we can use the <code>apply_filter</code> function to effectively apply filters. The strings passed to the function represent expressions that will be evaluated, thus keeping only PSMs that have 0 irregular cleavages and 2 or less missed cleavages.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="st">"numIrregCleavages == 0"</span><span class="op">)</span></span>
<span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="st">"numMissCleavages &lt;= 2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="co">##  MSnID object</span></span>
<span><span class="co">##  Working directory: "."</span></span>
<span><span class="co">##  #Spectrum Files:  1 </span></span>
<span><span class="co">##  #PSMs: 7838 at 17 % FDR</span></span>
<span><span class="co">##  #peptides: 5598 at 23 % FDR</span></span>
<span><span class="co">##  #accessions: 3759 at 53 % FDR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="parent-ion-mass-errors" class="level3" data-number="3.9.4"><h3 data-number="3.9.4" class="anchored" data-anchor-id="parent-ion-mass-errors">
<span class="header-section-number">3.9.4</span> Parent ion mass errors</h3>
<p>Using <code>"calculatedMassToCharge"</code> and <code>"experimentalMassToCharge"</code>, the <code>mass_measurement_error</code> function calculates the parent ion mass measurement error in parts per million.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/mass_measurement_error.html">mass_measurement_error</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##        Min.    1st Qu.     Median       Mean    3rd Qu.       Max. </span></span>
<span><span class="co">##  -2184.0640    -0.6992     0.0000    17.6146     0.7512  2012.5178</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then filter any matches that do not fit the +/- 20 ppm tolerance</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="st">"abs(mass_measurement_error(msnid)) &lt; 20"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/mass_measurement_error.html">mass_measurement_error</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##      Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">##  -19.7797  -0.5866   0.0000  -0.2970   0.5713  19.6758</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="filtering-criteria" class="level3" data-number="3.9.5"><h3 data-number="3.9.5" class="anchored" data-anchor-id="filtering-criteria">
<span class="header-section-number">3.9.5</span> Filtering criteria</h3>
<p>Filtering of the identification data will rely on</p>
<ul>
<li>-log10 transformed MS-GF+ Spectrum E-value, reflecting the goodness of match between experimental and theoretical fragmentation patterns</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span><span class="op">$</span><span class="va">msmsScore</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">$</span><span class="va">`MS-GF:SpecEValue`</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>the absolute mass measurement error (in ppm units) of the parent ion</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span><span class="op">$</span><span class="va">absParentMassErrorPPM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/mass_measurement_error.html">mass_measurement_error</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="setting-filters" class="level3" data-number="3.9.6"><h3 data-number="3.9.6" class="anchored" data-anchor-id="setting-filters">
<span class="header-section-number">3.9.6</span> Setting filters</h3>
<p>MS2 filters are handled by a special <code>MSnIDFilter</code> class objects, where individual filters are set by name (that is present in <code>names(msnid)</code>) and comparison operator (&gt;, &lt;, = , …) defining if we should retain hits with higher or lower given the threshold and finally the threshold value itself.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">filtObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/MSnIDFilter-class.html">MSnIDFilter</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="va">filtObj</span><span class="op">$</span><span class="va">absParentMassErrorPPM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>comparison<span class="op">=</span><span class="st">"&lt;"</span>, threshold<span class="op">=</span><span class="fl">10.0</span><span class="op">)</span></span>
<span><span class="va">filtObj</span><span class="op">$</span><span class="va">msmsScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>comparison<span class="op">=</span><span class="st">"&gt;"</span>, threshold<span class="op">=</span><span class="fl">10.0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">filtObj</span><span class="op">)</span></span>
<span><span class="co">##  MSnIDFilter object</span></span>
<span><span class="co">##  (absParentMassErrorPPM &lt; 10) &amp; (msmsScore &gt; 10)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then evaluate the filter on the identification data object, which returns the false discovery rate and number of retained identifications for the filtering criteria at hand.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/evaluate_filter.html">evaluate_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="va">filtObj</span><span class="op">)</span></span>
<span><span class="co">##            fdr    n</span></span>
<span><span class="co">##  PSM         0 3807</span></span>
<span><span class="co">##  peptide     0 2455</span></span>
<span><span class="co">##  accession   0 1009</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="filter-optimisation" class="level3" data-number="3.9.7"><h3 data-number="3.9.7" class="anchored" data-anchor-id="filter-optimisation">
<span class="header-section-number">3.9.7</span> Filter optimisation</h3>
<p>Rather than setting filtering values by hand, as shown above, these can be set automatically to meet a specific false discovery rate.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">filtObj.grid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/optimize_filter.html">optimize_filter</a></span><span class="op">(</span><span class="va">filtObj</span>, <span class="va">msnid</span>, fdr.max<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>                                method<span class="op">=</span><span class="st">"Grid"</span>, level<span class="op">=</span><span class="st">"peptide"</span>,</span>
<span>                                n.iter<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">filtObj.grid</span><span class="op">)</span></span>
<span><span class="co">##  MSnIDFilter object</span></span>
<span><span class="co">##  (absParentMassErrorPPM &lt; 3) &amp; (msmsScore &gt; 7.4)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/evaluate_filter.html">evaluate_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="va">filtObj.grid</span><span class="op">)</span></span>
<span><span class="co">##                    fdr    n</span></span>
<span><span class="co">##  PSM       0.004097561 5146</span></span>
<span><span class="co">##  peptide   0.006447651 3278</span></span>
<span><span class="co">##  accession 0.021996616 1208</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Filters can eventually be applied (rather than just evaluated) using the <code>apply_filter</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="va">filtObj.grid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="co">##  MSnID object</span></span>
<span><span class="co">##  Working directory: "."</span></span>
<span><span class="co">##  #Spectrum Files:  1 </span></span>
<span><span class="co">##  #PSMs: 5146 at 0.41 % FDR</span></span>
<span><span class="co">##  #peptides: 3278 at 0.64 % FDR</span></span>
<span><span class="co">##  #accessions: 1208 at 2.2 % FDR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And finally, identifications that matched decoy and contaminant protein sequences are removed</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="st">"isDecoy == FALSE"</span><span class="op">)</span></span>
<span><span class="va">msnid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnID/man/apply_filter.html">apply_filter</a></span><span class="op">(</span><span class="va">msnid</span>, <span class="st">"!grepl('Contaminant',accession)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html">show</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span></span>
<span><span class="co">##  MSnID object</span></span>
<span><span class="co">##  Working directory: "."</span></span>
<span><span class="co">##  #Spectrum Files:  1 </span></span>
<span><span class="co">##  #PSMs: 5117 at 0 % FDR</span></span>
<span><span class="co">##  #peptides: 3251 at 0 % FDR</span></span>
<span><span class="co">##  #accessions: 1179 at 0 % FDR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="export-msnid-data" class="level3" data-number="3.9.8"><h3 data-number="3.9.8" class="anchored" data-anchor-id="export-msnid-data">
<span class="header-section-number">3.9.8</span> Export <code>MSnID</code> data</h3>
<p>The resulting filtered identification data can be exported to a <code>data.frame</code> (or to a dedicated <code>MSnSet</code> data structure from the <code>MSnbase</code> package) for quantitative MS data, described below, and further processed and analysed using appropriate statistical tests.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">psms</a></span><span class="op">(</span><span class="va">msnid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##    spectrumID scan number(s) acquisitionNum passThreshold rank</span></span>
<span><span class="co">##  1 index=7151           8819           7151          TRUE    1</span></span>
<span><span class="co">##  2 index=8520          10419           8520          TRUE    1</span></span>
<span><span class="co">##  3 index=6683           8279           6683          TRUE    1</span></span>
<span><span class="co">##  4 index=6683           8279           6683          TRUE    1</span></span>
<span><span class="co">##  5 index=6683           8279           6683          TRUE    1</span></span>
<span><span class="co">##  6 index=6683           8279           6683          TRUE    1</span></span>
<span><span class="co">##    calculatedMassToCharge experimentalMassToCharge chargeState</span></span>
<span><span class="co">##  1               1270.318                 1270.318           3</span></span>
<span><span class="co">##  2               1426.737                 1426.739           3</span></span>
<span><span class="co">##  3               1440.348                 1440.350           3</span></span>
<span><span class="co">##  4               1440.348                 1440.350           3</span></span>
<span><span class="co">##  5               1440.348                 1440.350           3</span></span>
<span><span class="co">##  6               1440.348                 1440.350           3</span></span>
<span><span class="co">##    MS-GF:DeNovoScore MS-GF:EValue MS-GF:PepQValue MS-GF:QValue MS-GF:RawScore</span></span>
<span><span class="co">##  1               287 1.709082e-24               0            0            239</span></span>
<span><span class="co">##  2               270 3.780745e-24               0            0            230</span></span>
<span><span class="co">##  3               294 1.106378e-23               0            0            224</span></span>
<span><span class="co">##  4               294 1.106378e-23               0            0            224</span></span>
<span><span class="co">##  5               294 1.106378e-23               0            0            224</span></span>
<span><span class="co">##  6               294 1.106378e-23               0            0            224</span></span>
<span><span class="co">##    MS-GF:SpecEValue AssumedDissociationMethod IsotopeError isDecoy post pre</span></span>
<span><span class="co">##  1     1.007452e-31                       CID            0   FALSE    A   K</span></span>
<span><span class="co">##  2     2.217275e-31                       CID            0   FALSE    A   K</span></span>
<span><span class="co">##  3     6.504763e-31                       CID            0   FALSE    W   K</span></span>
<span><span class="co">##  4     6.504763e-31                       CID            0   FALSE    W   K</span></span>
<span><span class="co">##  5     6.504763e-31                       CID            0   FALSE    W   K</span></span>
<span><span class="co">##  6     6.504763e-31                       CID            0   FALSE    W   K</span></span>
<span><span class="co">##    end start accession length</span></span>
<span><span class="co">##  1 283   249   CE02347    393</span></span>
<span><span class="co">##  2 182   142   CE07055    206</span></span>
<span><span class="co">##  3 422   385   CE12728    654</span></span>
<span><span class="co">##  4 422   385   CE36358    582</span></span>
<span><span class="co">##  5 355   318   CE36359    587</span></span>
<span><span class="co">##  6 386   349   CE36360    618</span></span>
<span><span class="co">##                                                                                                                            description</span></span>
<span><span class="co">##  1 WBGene00001993; locus:hpd-1; 4-hydroxyphenylpyruvate dioxygenase; status:Confirmed; UniProt:Q22633; protein_id:CAA90315.1; T21C12.2</span></span>
<span><span class="co">##  2           WBGene00001755; locus:gst-7; glutathione S-transferase; status:Confirmed; UniProt:P91253; protein_id:AAB37846.1; F11G11.2</span></span>
<span><span class="co">##  3                WBGene00011232; phosphoenolpyruvate carboxykinase; status:Confirmed; UniProt:O02286; protein_id:CAB05600.1; R11A5.4a</span></span>
<span><span class="co">##  4                                         WBGene00011232; status:Partially_confirmed; UniProt:Q7JKI1; protein_id:CAF31484.1; R11A5.4b</span></span>
<span><span class="co">##  5                                                   WBGene00011232; status:Confirmed; UniProt:Q7JKI3; protein_id:CAF31482.1; R11A5.4c</span></span>
<span><span class="co">##  6                                                   WBGene00011232; status:Confirmed; UniProt:Q7JKI2; protein_id:CAF31483.1; R11A5.4d</span></span>
<span><span class="co">##                                       pepSeq modified modification</span></span>
<span><span class="co">##  1       AISQIQEYVDYYGGSGVQHIALNTSDIITAIEALR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  2 SAGSGYLVGDSLTFVDLLVAQHTADLLAANAALLDEFPQFK    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  3    NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  4    NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  5    NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  6    NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##               idFile                                   spectrumFile</span></span>
<span><span class="co">##  1 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##  2 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##  3 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##  4 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##  5 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##  6 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt</span></span>
<span><span class="co">##                databaseFile                                       peptide</span></span>
<span><span class="co">##  1 ID_004174_E48C5B52.fasta       K.AISQIQEYVDYYGGSGVQHIALNTSDIITAIEALR.A</span></span>
<span><span class="co">##  2 ID_004174_E48C5B52.fasta K.SAGSGYLVGDSLTFVDLLVAQHTADLLAANAALLDEFPQFK.A</span></span>
<span><span class="co">##  3 ID_004174_E48C5B52.fasta    K.NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK.W</span></span>
<span><span class="co">##  4 ID_004174_E48C5B52.fasta    K.NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK.W</span></span>
<span><span class="co">##  5 ID_004174_E48C5B52.fasta    K.NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK.W</span></span>
<span><span class="co">##  6 ID_004174_E48C5B52.fasta    K.NSIFTNVAETANGEYFWEGLEDEIADKNVDITTWLGEK.W</span></span>
<span><span class="co">##    numIrregCleavages numMissCleavages msmsScore absParentMassErrorPPM</span></span>
<span><span class="co">##  1                 0                0  30.99678             0.3843772</span></span>
<span><span class="co">##  2                 0                0  30.65418             1.3689451</span></span>
<span><span class="co">##  3                 0                1  30.18677             0.9322561</span></span>
<span><span class="co">##  4                 0                1  30.18677             0.9322561</span></span>
<span><span class="co">##  5                 0                1  30.18677             0.9322561</span></span>
<span><span class="co">##  6                 0                1  30.18677             0.9322561</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Morgenstern:2020" class="csl-entry" role="listitem">
Morgenstern, David, Rotem Barzilay, and Yishai Levin. 2021. <span>“<span>RawBeans</span>: A Simple, Vendor-Independent, Raw-Data Quality-Control Tool.”</span> <em>Journal of Proteome Research</em>. <a href="https://doi.org/10.1021/acs.jproteome.0c00956">https://doi.org/10.1021/acs.jproteome.0c00956</a>.
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>Previously named <code>PSM</code>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The rownames aren’t needed here are are removed to reduce to output in the the next code chunk display parts of <code>id2</code>.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/raw-ms-data.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Raw MS data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/30-quant.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>