<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R for Mass Spectrometry - 4&nbsp; Quantitative data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/95-annex.html" rel="next">
<link href="../pages/20-id.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/30-quant.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">R for Mass Spectrometry</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/R4MS/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/R4MS/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/raw-ms-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Raw MS data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20-id.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30-quant.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95-annex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/99-si.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Additional materials and help</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#quantitation-methodologies" id="toc-quantitation-methodologies" class="nav-link active" data-scroll-target="#quantitation-methodologies"><span class="header-section-number">4.1</span> Quantitation methodologies</a>
  <ul class="collapse">
<li><a href="#label-free-ms2-spectral-counting" id="toc-label-free-ms2-spectral-counting" class="nav-link" data-scroll-target="#label-free-ms2-spectral-counting"><span class="header-section-number">4.1.1</span> Label-free MS2: Spectral counting</a></li>
  <li><a href="#labelled-ms2-isobaric-tagging" id="toc-labelled-ms2-isobaric-tagging" class="nav-link" data-scroll-target="#labelled-ms2-isobaric-tagging"><span class="header-section-number">4.1.2</span> Labelled MS2: Isobaric tagging</a></li>
  <li><a href="#label-free-ms1-extracted-ion-chromatograms" id="toc-label-free-ms1-extracted-ion-chromatograms" class="nav-link" data-scroll-target="#label-free-ms1-extracted-ion-chromatograms"><span class="header-section-number">4.1.3</span> Label-free MS1: extracted ion chromatograms</a></li>
  <li><a href="#labelled-ms1-silac" id="toc-labelled-ms1-silac" class="nav-link" data-scroll-target="#labelled-ms1-silac"><span class="header-section-number">4.1.4</span> Labelled MS1: SILAC</a></li>
  </ul>
</li>
  <li>
<a href="#sec-qf" id="toc-sec-qf" class="nav-link" data-scroll-target="#sec-qf"><span class="header-section-number">4.2</span> QFeatures</a>
  <ul class="collapse">
<li><a href="#the-qfeatures-class" id="toc-the-qfeatures-class" class="nav-link" data-scroll-target="#the-qfeatures-class"><span class="header-section-number">4.2.1</span> The QFeatures class</a></li>
  <li><a href="#feature-aggregation" id="toc-feature-aggregation" class="nav-link" data-scroll-target="#feature-aggregation"><span class="header-section-number">4.2.2</span> Feature aggregation</a></li>
  <li><a href="#subsetting-and-filtering" id="toc-subsetting-and-filtering" class="nav-link" data-scroll-target="#subsetting-and-filtering"><span class="header-section-number">4.2.3</span> Subsetting and filtering</a></li>
  </ul>
</li>
  <li>
<a href="#creating-qfeatures-object" id="toc-creating-qfeatures-object" class="nav-link" data-scroll-target="#creating-qfeatures-object"><span class="header-section-number">4.3</span> Creating <code>QFeatures</code> object</a>
  <ul class="collapse">
<li><a href="#exercise" id="toc-exercise" class="nav-link" data-scroll-target="#exercise">Exercise</a></li>
  </ul>
</li>
  <li>
<a href="#analysis-pipeline" id="toc-analysis-pipeline" class="nav-link" data-scroll-target="#analysis-pipeline"><span class="header-section-number">4.4</span> Analysis pipeline</a>
  <ul class="collapse">
<li><a href="#missing-values" id="toc-missing-values" class="nav-link" data-scroll-target="#missing-values"><span class="header-section-number">4.4.1</span> Missing values</a></li>
  <li><a href="#imputation" id="toc-imputation" class="nav-link" data-scroll-target="#imputation"><span class="header-section-number">4.4.2</span> Imputation</a></li>
  <li><a href="#identification-quality-control" id="toc-identification-quality-control" class="nav-link" data-scroll-target="#identification-quality-control"><span class="header-section-number">4.4.3</span> Identification quality control</a></li>
  <li><a href="#creating-the-qfeatures-data" id="toc-creating-the-qfeatures-data" class="nav-link" data-scroll-target="#creating-the-qfeatures-data"><span class="header-section-number">4.4.4</span> Creating the QFeatures data</a></li>
  <li><a href="#filtering-out-contaminants-and-reverse-hits" id="toc-filtering-out-contaminants-and-reverse-hits" class="nav-link" data-scroll-target="#filtering-out-contaminants-and-reverse-hits"><span class="header-section-number">4.4.5</span> Filtering out contaminants and reverse hits</a></li>
  <li><a href="#log-transformation-and-normalisation" id="toc-log-transformation-and-normalisation" class="nav-link" data-scroll-target="#log-transformation-and-normalisation"><span class="header-section-number">4.4.6</span> Log-transformation and normalisation</a></li>
  <li><a href="#aggregation" id="toc-aggregation" class="nav-link" data-scroll-target="#aggregation"><span class="header-section-number">4.4.7</span> Aggregation</a></li>
  <li><a href="#principal-component-analysis" id="toc-principal-component-analysis" class="nav-link" data-scroll-target="#principal-component-analysis"><span class="header-section-number">4.4.8</span> Principal component analysis</a></li>
  <li><a href="#visualisation" id="toc-visualisation" class="nav-link" data-scroll-target="#visualisation"><span class="header-section-number">4.4.9</span> Visualisation</a></li>
  <li><a href="#statistical-analysis" id="toc-statistical-analysis" class="nav-link" data-scroll-target="#statistical-analysis"><span class="header-section-number">4.4.10</span> Statistical analysis</a></li>
  </ul>
</li>
  <li><a href="#summary-exercice" id="toc-summary-exercice" class="nav-link" data-scroll-target="#summary-exercice"><span class="header-section-number">4.5</span> Summary exercice</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-quant" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><section id="quantitation-methodologies" class="level2" data-number="4.1"><h2 data-number="4.1" class="anchored" data-anchor-id="quantitation-methodologies">
<span class="header-section-number">4.1</span> Quantitation methodologies</h2>
<p>There are a wide range of proteomics quantitation techniques that can broadly be classified as labelled vs.&nbsp;label-free, depending on whether the features are labelled prior the MS acquisition and the MS level at which quantitation is inferred, namely MS1 or MS2.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Label-free</th>
<th style="text-align: left;">Labelled</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">MS1</td>
<td style="text-align: left;">XIC</td>
<td style="text-align: left;">SILAC, 15N</td>
</tr>
<tr class="even">
<td style="text-align: left;">MS2</td>
<td style="text-align: left;">Counting</td>
<td style="text-align: left;">iTRAQ, TMT</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="label-free-ms2-spectral-counting" class="level3" data-number="4.1.1"><h3 data-number="4.1.1" class="anchored" data-anchor-id="label-free-ms2-spectral-counting">
<span class="header-section-number">4.1.1</span> Label-free MS2: Spectral counting</h3>
<p>In spectral counting, one simply counts the number of quantified peptides that are assigned to a protein.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/pbase.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption">Spectral counting. Figure from the <code>Pbase</code> package.</figcaption></figure>
</div>
</div>
</div>
</section><section id="labelled-ms2-isobaric-tagging" class="level3" data-number="4.1.2"><h3 data-number="4.1.2" class="anchored" data-anchor-id="labelled-ms2-isobaric-tagging">
<span class="header-section-number">4.1.2</span> Labelled MS2: Isobaric tagging</h3>
<p>Isobaric tagging refers to the labelling using isobaric tags, i.e.&nbsp;chemical tags that have the same mass and hence can’t be distinguished by the spectrometer. The peptides of different samples (4, 6, 10, 11 or 16) are labelled with different tags and combined prior to mass spectrometry acquisition. Given that they are isobaric, all identical peptides, irrespective of the tag and this the sample of origin, are co-analysed, up to fragmentation prior to MS2 analysis. During fragmentation, the isobaric tags fall of, fragment themselves, and result in a set of sample specific peaks. These specific peaks can be used to infer sample-specific quantitation, while the rest of the MS2 spectrum is used for identification.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/itraq.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption">iTRAQ 4-plex isobaric tagging. Tandem Mass Tags (TMT) offer up to 16 tags.</figcaption></figure>
</div>
</div>
</div>
</section><section id="label-free-ms1-extracted-ion-chromatograms" class="level3" data-number="4.1.3"><h3 data-number="4.1.3" class="anchored" data-anchor-id="label-free-ms1-extracted-ion-chromatograms">
<span class="header-section-number">4.1.3</span> Label-free MS1: extracted ion chromatograms</h3>
<p>In label-free quantitation, the precursor peaks that match an identified peptide are integrated over retention time and the area under that <em>extracted ion chromatogram</em> is used to quantify that peptide in that sample.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/chrompeaks.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption">Label-free quantitation. Figure credit <a href="https://github.com/jorainer/">Johannes Rainer</a>.</figcaption></figure>
</div>
</div>
</div>
</section><section id="labelled-ms1-silac" class="level3" data-number="4.1.4"><h3 data-number="4.1.4" class="anchored" data-anchor-id="labelled-ms1-silac">
<span class="header-section-number">4.1.4</span> Labelled MS1: SILAC</h3>
<p>In SILAC quantitation, sample are grown in a medium that contains heavy amino acids (typically arginine and lysine). All proteins grown in this <em>heavy</em> growth medium contain the heavy form of these amino acids. Two samples, one grown in heavy medium, and one grown in normal (light) medium are then combined and analysed together. The heavy peptides precursor peaks are systematically shifted compared to the light ones, and the ratio between the height of a heavy and light peaks can be used to calculate peptide and protein fold-changes.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/Silac.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption">Silac quantitation. Figure credit Wikimedia Commons.</figcaption></figure>
</div>
</div>
</div>
<p>These different quantitation techniques come with their respective benefits and distinct challenges, such as large quantities of raw data processing, data transformation and normalisation, missing values, and different underlying statistical models for the quantitative data (count data for spectral counting, continuous data for the others).</p>
<p>In terms of raw data quantitation in R/Bioconductor, most efforts have been devoted to MS2-level quantitation. Label-free XIC quantitation has been addressed in the frame of metabolomics data processing by the <em><a href="https://bioconductor.org/packages/3.19/xcms">xcms</a></em> infrastructure.</p>
<!-- Below is a list of suggested packages for some common proteomics -->
<!-- quantitation technologies: -->
<!-- * Isobaric tagging (iTRAQ and TMT): *[MSnbase](https://bioconductor.org/packages/3.19/MSnbase)* and *[isobar](https://bioconductor.org/packages/3.19/isobar)*. -->
<!-- * Label-free: *[xcms](https://bioconductor.org/packages/3.19/xcms)* (metabolomics). -->
<!-- * Counting: *[MSnbase](https://bioconductor.org/packages/3.19/MSnbase)* and *[MSnID](https://bioconductor.org/packages/3.19/MSnID)* for -->
<!--   peptide-spectrum matching confidence assessment. -->
<!-- * *[N14N15](https://github.com/vladpetyuk/N14N15)* for heavy Nitrogen-labelled data. -->
</section></section><section id="sec-qf" class="level2" data-number="4.2"><h2 data-number="4.2" class="anchored" data-anchor-id="sec-qf">
<span class="header-section-number">4.2</span> QFeatures</h2>
<p>Mass spectrometry-based quantitative proteomics data can be represented as a matrix of quantitative values for features (PSMs, peptides, proteins) arranged along the rows, measured for a set of samples, arranged along the columns. There is a common representation for such quantitative data set, namely the <code>SummarizedExperiment</code> <span class="citation" data-cites="SE">(<a href="#ref-SE" role="doc-biblioref">Morgan et al. 2020</a>)</span> class:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/SE.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Schematic representation of the anatomy of a <code>SummarizedExperiment</code> object. (Figure taken from the <code>SummarizedExperiment</code> package vignette.)</figcaption></figure>
</div>
</div>
</div>
<ul>
<li>The sample (columns) metadata can be accessed with the <code>colData()</code> function.</li>
<li>The features (rows) metadata can be accessed with the <code>rowData()</code> column.</li>
<li>If the features represent ranges along genomic coordinates, these can be accessed with <code>rowRanges()</code>
</li>
<li>Additional metadata describing the overall experiment can be accessed with <code>metadata()</code>.</li>
<li>The quantitative data can be accessed with <code>assay()</code>.</li>
<li>
<code>assays()</code> returns a list of matrix-like assays.</li>
</ul>
<section id="the-qfeatures-class" class="level3" data-number="4.2.1"><h3 data-number="4.2.1" class="anchored" data-anchor-id="the-qfeatures-class">
<span class="header-section-number">4.2.1</span> The QFeatures class</h3>
<p>While mass spectrometers acquire data for spectra/peptides, the biological entity of interest are the protein. As part of the data processing, we are thus required to <strong>aggregate</strong> low-level quantitative features into higher level data.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/featuresplot-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Conceptual representation of a <code>QFeatures</code> object and the aggregative relation between different assays.</figcaption></figure>
</div>
</div>
</div>
<p>We are going to start to familiarise ourselves with the <code>QFeatures</code> class implemented in the <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html"><code>QFeatures</code></a> package. The class is derived from the Bioconductor <code>MultiAssayExperiment</code> <span class="citation" data-cites="MAE">(<a href="#ref-MAE" role="doc-biblioref">Ramos et al. 2017</a>)</span> (MAE) class. Let’s start by loading the <code>QFeatures</code> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures">QFeatures</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we load the <code>feat1</code> test data, which is composed of single <em>assay</em> of class <code>SummarizedExperiment</code> composed of 10 rows and 2 columns.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">feat1</span><span class="op">)</span></span>
<span><span class="va">feat1</span></span>
<span><span class="co">##  An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 10 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s perform some simple operations to familiarise ourselves with the <code>QFeatures</code> class:</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Extract the sample metadata using the <code>colData()</code> accessor (like you have previously done with <code>SummarizedExperiment</code> objects).</li>
</ul>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">feat1</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 2 rows and 1 column</span></span>
<span><span class="co">##         Group</span></span>
<span><span class="co">##     &lt;integer&gt;</span></span>
<span><span class="co">##  S1         1</span></span>
<span><span class="co">##  S2         2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>We can also further annotate the experiment by adding columns to the <code>colData</code> slot:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">feat1</span><span class="op">)</span><span class="op">$</span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"X1"</span>, <span class="st">"X2"</span><span class="op">)</span></span>
<span><span class="va">feat1</span><span class="op">$</span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Y1"</span>, <span class="st">"Y2"</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">feat1</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 2 rows and 3 columns</span></span>
<span><span class="co">##         Group           X           Y</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##  S1         1          X1          Y1</span></span>
<span><span class="co">##  S2         2          X2          Y2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Extract the first (and only) assay composing this <code>QFeatures</code> data using the <code>[[</code> operator (as you have done to extract elements of a list) by using the assay’s index or name.</li>
</ul>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feat1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 10 2 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(10): PSM1 PSM2 ... PSM9 PSM10</span></span>
<span><span class="co">##  rowData names(5): Sequence Protein Var location pval</span></span>
<span><span class="co">##  colnames(2): S1 S2</span></span>
<span><span class="co">##  colData names(0):</span></span>
<span><span class="va">feat1</span><span class="op">[[</span><span class="st">"psms"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 10 2 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(10): PSM1 PSM2 ... PSM9 PSM10</span></span>
<span><span class="co">##  rowData names(5): Sequence Protein Var location pval</span></span>
<span><span class="co">##  colnames(2): S1 S2</span></span>
<span><span class="co">##  colData names(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="feature-aggregation" class="level3" data-number="4.2.2"><h3 data-number="4.2.2" class="anchored" data-anchor-id="feature-aggregation">
<span class="header-section-number">4.2.2</span> Feature aggregation</h3>
<p>The central functionality of the <code>QFeatures</code> infrastructure is the aggregation of features into higher-level features while retaining the link between the different levels. This can be done with the <a href="https://rformassspectrometry.github.io/QFeatures/reference/QFeatures-aggregate.html"><code>aggregateFeatures()</code> function</a>.</p>
<p>The call below will</p>
<ul>
<li>operate on the <code>psms</code> assay of the <code>feat1</code> objects;</li>
<li>aggregate the rows of the assay following the grouping defined in the <code>peptides</code> row data variables;</li>
<li>perform aggregation using the <code><a href="https://rdrr.io/r/base/colSums.html">colMeans()</a></code> function;</li>
<li>create a new assay named <code>peptides</code> and add it to the <code>feat1</code> object.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">aggregateFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, i <span class="op">=</span> <span class="st">"psms"</span>,</span>
<span>                           fcol <span class="op">=</span> <span class="st">"Sequence"</span>,</span>
<span>                           name <span class="op">=</span> <span class="st">"peptides"</span>,</span>
<span>                           fun <span class="op">=</span> <span class="va">colMeans</span><span class="op">)</span></span>
<span><span class="va">feat1</span></span>
<span><span class="co">##  An instance of class QFeatures containing 2 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 10 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 3 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Let’s convince ourselves that we understand the effect of feature aggregation and repeat the calculations manually and check the content of the new assay’s row data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## SYGFNAAR</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  S1 S2 </span></span>
<span><span class="co">##   2 12</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="st">"SYGFNAAR"</span>, <span class="op">]</span></span>
<span><span class="co">##  S1 S2 </span></span>
<span><span class="co">##   2 12</span></span>
<span></span>
<span><span class="co">## ELGNDAYK</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">4</span><span class="op">:</span><span class="fl">6</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  S1 S2 </span></span>
<span><span class="co">##   5 15</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="st">"ELGNDAYK"</span>, <span class="op">]</span></span>
<span><span class="co">##  S1 S2 </span></span>
<span><span class="co">##   5 15</span></span>
<span></span>
<span><span class="co">## IAEESNFPFIK</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##    S1   S2 </span></span>
<span><span class="co">##   8.5 18.5</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="st">"IAEESNFPFIK"</span>, <span class="op">]</span></span>
<span><span class="co">##    S1   S2 </span></span>
<span><span class="co">##   8.5 18.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 3 rows and 4 columns</span></span>
<span><span class="co">##                   Sequence     Protein      location        .n</span></span>
<span><span class="co">##                &lt;character&gt; &lt;character&gt;   &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  ELGNDAYK         ELGNDAYK       ProtA Mitochondr...         3</span></span>
<span><span class="co">##  IAEESNFPFIK IAEESNFPFI...       ProtB       unknown         4</span></span>
<span><span class="co">##  SYGFNAAR         SYGFNAAR       ProtA Mitochondr...         3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now aggregate the peptide-level data into a new protein-level assay using the <code>colMedians()</code> aggregation function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">aggregateFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, i <span class="op">=</span> <span class="st">"peptides"</span>,</span>
<span>                           fcol <span class="op">=</span> <span class="st">"Protein"</span>,</span>
<span>                           name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                           fun <span class="op">=</span> <span class="va">colMedians</span><span class="op">)</span></span>
<span><span class="va">feat1</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 10 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 3 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 2 rows and 2 columns</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">feat1</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##         S1   S2</span></span>
<span><span class="co">##  ProtA 3.5 13.5</span></span>
<span><span class="co">##  ProtB 8.5 18.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="subsetting-and-filtering" class="level3" data-number="4.2.3"><h3 data-number="4.2.3" class="anchored" data-anchor-id="subsetting-and-filtering">
<span class="header-section-number">4.2.3</span> Subsetting and filtering</h3>
<p>The link between the assays becomes apparent when we now subset the assays for protein A as shown below or using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-subsetBy.html">subsetByFeature()</a></code> function. This creates a new instance of class <code>QFeatures</code> containing assays with the expression data for protein, its peptides and their PSMs.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feat1</span><span class="op">[</span><span class="st">"ProtA"</span>, , <span class="op">]</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 6 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 2 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 1 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures()</a></code> function can be used to filter rows the assays composing a <code>QFeatures</code> object using the row data variables. We can for example retain rows that have a <code>pval</code> &lt; 0.05, which would only keep rows in the <code>psms</code> assay because the <code>pval</code> is only relevant for that assay.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, <span class="op">~</span> <span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">##  'pval' found in 1 out of 3 assay(s)</span></span>
<span><span class="co">##  No filter applied to the following assay(s) because one or more filtering variables are missing in the rowData: peptides, proteins.</span></span>
<span><span class="co">##  You can control whether to remove or keep the features using the 'keep' argument (see '?filterFeature').</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 4 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 0 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 0 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>As the message above implies, it is also possible to apply a filter to only the assays that have a filtering variables by setting the <code>keep</code> variables.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, <span class="op">~</span> <span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  'pval' found in 1 out of 3 assay(s)</span></span>
<span><span class="co">##  No filter applied to the following assay(s) because one or more filtering variables are missing in the rowData: peptides, proteins.</span></span>
<span><span class="co">##  You can control whether to remove or keep the features using the 'keep' argument (see '?filterFeature').</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 4 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 3 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 2 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>On the other hand, if we filter assay rows for those that localise to the mitochondrion, we retain the relevant protein, peptides and PSMs.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, <span class="op">~</span> <span class="va">location</span> <span class="op">==</span> <span class="st">"Mitochondrion"</span><span class="op">)</span></span>
<span><span class="co">##  'location' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 6 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 2 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 1 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>As an exercise, let’s filter rows that do not localise to the mitochondrion.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">feat1</span>, <span class="op">~</span> <span class="va">location</span> <span class="op">!=</span> <span class="st">"Mitochondrion"</span><span class="op">)</span></span>
<span><span class="co">##  'location' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">##  An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 4 rows and 2 columns </span></span>
<span><span class="co">##   [2] peptides: SummarizedExperiment with 1 rows and 2 columns </span></span>
<span><span class="co">##   [3] proteins: SummarizedExperiment with 1 rows and 2 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>You can refer to the <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html"><em>Quantitative features for mass spectrometry data</em></a> vignette and the <code>QFeatures</code> <a href="https://rformassspectrometry.github.io/QFeatures/reference/QFeatures-class.html">manual page</a> for more details about the class.</p>
</section></section><section id="creating-qfeatures-object" class="level2" data-number="4.3"><h2 data-number="4.3" class="anchored" data-anchor-id="creating-qfeatures-object">
<span class="header-section-number">4.3</span> Creating <code>QFeatures</code> object</h2>
<p>While <code>QFeatures</code> objects can be created manually (see <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">?QFeatures</a></code> for details), most users have a quantitative data in a spreadsheet or a data.frame. In such cases, the easiest is to use the <code>readQFeatures</code> function to extract the quantitative data and metadata columns. Below, we load the <code>hlpsms</code> dataframe that contains data for 28 PSMs from the TMT-10plex <em>hyper</em>LOPIT spatial proteomics experiment from <span class="citation" data-cites="Christoforou:2016">(<a href="#ref-Christoforou:2016" role="doc-biblioref">Christoforou et al. 2016</a>)</span>. The <code>ecol</code> argument specifies that columns 1 to 10 contain quantitation data, and that the assay should be named <code>psms</code> in the returned <code>QFeatures</code> object, to reflect the nature of the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">hlpsms</span><span class="op">)</span></span>
<span><span class="va">hl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readQFeatures</a></span><span class="op">(</span><span class="va">hlpsms</span>, ecol <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, name <span class="op">=</span> <span class="st">"psms"</span><span class="op">)</span></span>
<span><span class="va">hl</span></span>
<span><span class="co">##  An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">##   [1] psms: SummarizedExperiment with 3010 rows and 10 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Below, we see that we can extract an assay using its index or its name. The individual assays are stored as <em>SummarizedExperiment</em> object and further access its quantitative data and metadata using the <code>assay</code> and <code>rowData</code> functions.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 3010 10 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(3010): 1 2 ... 3009 3010</span></span>
<span><span class="co">##  rowData names(18): Sequence ProteinDescriptions ... RTmin markers</span></span>
<span><span class="co">##  colnames(10): X126 X127C ... X130N X131</span></span>
<span><span class="co">##  colData names(0):</span></span>
<span><span class="va">hl</span><span class="op">[[</span><span class="st">"psms"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 3010 10 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(3010): 1 2 ... 3009 3010</span></span>
<span><span class="co">##  rowData names(18): Sequence ProteinDescriptions ... RTmin markers</span></span>
<span><span class="co">##  colnames(10): X126 X127C ... X130N X131</span></span>
<span><span class="co">##  colData names(0):</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">hl</span><span class="op">[[</span><span class="st">"psms"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          X126      X127C       X127N      X128C       X128N      X129C</span></span>
<span><span class="co">##  1 0.12283431 0.08045915 0.070804055 0.09386901 0.051815695 0.13034383</span></span>
<span><span class="co">##  2 0.35268185 0.14162381 0.167523880 0.07843497 0.071087436 0.03214548</span></span>
<span><span class="co">##  3 0.01546089 0.16142297 0.086938133 0.23120844 0.114664348 0.09610188</span></span>
<span><span class="co">##  4 0.04702854 0.09288723 0.102012167 0.11125409 0.067969116 0.14155358</span></span>
<span><span class="co">##  5 0.01044693 0.15866147 0.167315736 0.21017494 0.147946673 0.07088253</span></span>
<span><span class="co">##  6 0.04955362 0.01215244 0.002477681 0.01297833 0.002988949 0.06253195</span></span>
<span><span class="co">##         X129N       X130C      X130N       X131</span></span>
<span><span class="co">##  1 0.17540095 0.040068658 0.11478839 0.11961594</span></span>
<span><span class="co">##  2 0.06686260 0.031961793 0.02810434 0.02957384</span></span>
<span><span class="co">##  3 0.15977819 0.010127118 0.08059400 0.04370403</span></span>
<span><span class="co">##  4 0.18015910 0.035329902 0.12166589 0.10014038</span></span>
<span><span class="co">##  5 0.17555789 0.007088253 0.02884754 0.02307803</span></span>
<span><span class="co">##  6 0.01726511 0.172651119 0.37007905 0.29732174</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">hl</span><span class="op">[[</span><span class="st">"psms"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 18 columns</span></span>
<span><span class="co">##       Sequence ProteinDescriptions NbProteins ProteinGroupAccessions</span></span>
<span><span class="co">##    &lt;character&gt;         &lt;character&gt;  &lt;integer&gt;            &lt;character&gt;</span></span>
<span><span class="co">##  1     SQGEIDk       Tetratrico...          1                 Q8BYY4</span></span>
<span><span class="co">##  2     YEAQGDk       Vacuolar p...          1                 P46467</span></span>
<span><span class="co">##  3     TTScDTk       C-type man...          1                 Q64449</span></span>
<span><span class="co">##  4     aEELESR       Liprin-alp...          1                 P60469</span></span>
<span><span class="co">##  5     aQEEAIk       Isoform 2 ...          2               P13597-2</span></span>
<span><span class="co">##  6    dGAVDGcR       Structural...          1                 Q6P5D8</span></span>
<span><span class="co">##    Modifications    qValue       PEP  IonScore NbMissedCleavages</span></span>
<span><span class="co">##      &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;integer&gt;</span></span>
<span><span class="co">##  1 K7(TMT6ple...     0.008   0.11800        27                 0</span></span>
<span><span class="co">##  2 K7(TMT6ple...     0.001   0.01070        27                 0</span></span>
<span><span class="co">##  3 C4(Carbami...     0.008   0.11800        11                 0</span></span>
<span><span class="co">##  4 N-Term(TMT...     0.002   0.04450        24                 0</span></span>
<span><span class="co">##  5 N-Term(Car...     0.001   0.00850        36                 0</span></span>
<span><span class="co">##  6 N-Term(TMT...     0.000   0.00322        26                 0</span></span>
<span><span class="co">##    IsolationInterference IonInjectTimems Intensity    Charge      mzDa</span></span>
<span><span class="co">##                &lt;integer&gt;       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  1                     0              70    335000         2   503.274</span></span>
<span><span class="co">##  2                     0              70    926000         2   520.267</span></span>
<span><span class="co">##  3                     0              70    159000         2   521.258</span></span>
<span><span class="co">##  4                     0              70    232000         2   531.785</span></span>
<span><span class="co">##  5                     0              70    212000         2   537.804</span></span>
<span><span class="co">##  6                     0              70    865000         2   539.761</span></span>
<span><span class="co">##         MHDa DeltaMassPPM     RTmin       markers</span></span>
<span><span class="co">##    &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">##  1   1005.54        -0.38     24.02       unknown</span></span>
<span><span class="co">##  2   1039.53         0.61     18.85       unknown</span></span>
<span><span class="co">##  3   1041.51         1.11     10.17       unknown</span></span>
<span><span class="co">##  4   1062.56         0.35     29.18       unknown</span></span>
<span><span class="co">##  5   1074.60         1.70     25.56 Plasma mem...</span></span>
<span><span class="co">##  6   1078.51        -0.67     21.27 Nucleus - ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For further details on how to manipulate such objects, refer to the <em><a href="https://bioconductor.org/packages/3.19/MultiAssayExperiment">MultiAssayExperiment</a></em> <span class="citation" data-cites="MAE">(<a href="#ref-MAE" role="doc-biblioref">Ramos et al. 2017</a>)</span> and <em><a href="https://bioconductor.org/packages/3.19/SummarizedExperiment">SummarizedExperiment</a></em> <span class="citation" data-cites="SE">(<a href="#ref-SE" role="doc-biblioref">Morgan et al. 2020</a>)</span> packages.</p>
<p>It is also possible to first create a <code>SummarizedExperiment</code>, and then only include it into a <code>QFeatures</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readSummarizedExperiment</a></span><span class="op">(</span><span class="va">hlpsms</span>, ecol <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">se</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 3010 10 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(3010): 1 2 ... 3009 3010</span></span>
<span><span class="co">##  rowData names(18): Sequence ProteinDescriptions ... RTmin markers</span></span>
<span><span class="co">##  colnames(10): X126 X127C ... X130N X131</span></span>
<span><span class="co">##  colData names(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">QFeatures</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>psm <span class="op">=</span> <span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">##   [1] psm: SummarizedExperiment with 3010 rows and 10 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>At this stage, i.e.&nbsp;at the beginning of the analysis, whether you have a <code>SummarizedExperiment</code> or a <code>QFeatures</code> object, it is a good time to define the experimental design in the <code>colData</code> slot.</p>
<section id="exercise" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="exercise">Exercise</h3>
<p>The CPTAC spike-in study 6 <span class="citation" data-cites="Paulovich:2010">(<a href="#ref-Paulovich:2010" role="doc-biblioref">Paulovich et al. 2010</a>)</span> combines the Sigma UPS1 standard containing 48 different human proteins that are spiked in at 5 different concentrations (conditions A to E) into a constant yeast protein background. The sample were acquired in triplicate on different instruments in different labs. We are going to start with a subset of the CPTAC study 6 containing conditions A and B for a single lab.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/cptac.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption class="figure-caption">The CPTAC spike-in study design (credit Lieven Clement, statOmics, Ghent University).</figcaption></figure>
</div>
</div>
</div>
<p>The peptide-level data, as processed by MaxQuant <span class="citation" data-cites="Cox:2008">(<a href="#ref-Cox:2008" role="doc-biblioref">Cox and Mann 2008</a>)</span> is available in the <code>msdata</code> package:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">quant</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"cptac"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "cptac_a_b_peptides.txt"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Read these data in as either a <code>SummarizedExperiment</code> or a <code>QFeatures</code> object.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>From the names of the columns, we see that the quantitative columns, starting with <code>"Intensity."</code> (note the dot!) are at positions 56 to 61.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##   [1] "Sequence"                 "N.term.cleavage.window"  </span></span>
<span><span class="co">##   [3] "C.term.cleavage.window"   "Amino.acid.before"       </span></span>
<span><span class="co">##   [5] "First.amino.acid"         "Second.amino.acid"       </span></span>
<span><span class="co">##   [7] "Second.last.amino.acid"   "Last.amino.acid"         </span></span>
<span><span class="co">##   [9] "Amino.acid.after"         "A.Count"                 </span></span>
<span><span class="co">##  [11] "R.Count"                  "N.Count"                 </span></span>
<span><span class="co">##  [13] "D.Count"                  "C.Count"                 </span></span>
<span><span class="co">##  [15] "Q.Count"                  "E.Count"                 </span></span>
<span><span class="co">##  [17] "G.Count"                  "H.Count"                 </span></span>
<span><span class="co">##  [19] "I.Count"                  "L.Count"                 </span></span>
<span><span class="co">##  [21] "K.Count"                  "M.Count"                 </span></span>
<span><span class="co">##  [23] "F.Count"                  "P.Count"                 </span></span>
<span><span class="co">##  [25] "S.Count"                  "T.Count"                 </span></span>
<span><span class="co">##  [27] "W.Count"                  "Y.Count"                 </span></span>
<span><span class="co">##  [29] "V.Count"                  "U.Count"                 </span></span>
<span><span class="co">##  [31] "Length"                   "Missed.cleavages"        </span></span>
<span><span class="co">##  [33] "Mass"                     "Proteins"                </span></span>
<span><span class="co">##  [35] "Leading.razor.protein"    "Start.position"          </span></span>
<span><span class="co">##  [37] "End.position"             "Unique..Groups."         </span></span>
<span><span class="co">##  [39] "Unique..Proteins."        "Charges"                 </span></span>
<span><span class="co">##  [41] "PEP"                      "Score"                   </span></span>
<span><span class="co">##  [43] "Identification.type.6A_7" "Identification.type.6A_8"</span></span>
<span><span class="co">##  [45] "Identification.type.6A_9" "Identification.type.6B_7"</span></span>
<span><span class="co">##  [47] "Identification.type.6B_8" "Identification.type.6B_9"</span></span>
<span><span class="co">##  [49] "Experiment.6A_7"          "Experiment.6A_8"         </span></span>
<span><span class="co">##  [51] "Experiment.6A_9"          "Experiment.6B_7"         </span></span>
<span><span class="co">##  [53] "Experiment.6B_8"          "Experiment.6B_9"         </span></span>
<span><span class="co">##  [55] "Intensity"                "Intensity.6A_7"          </span></span>
<span><span class="co">##  [57] "Intensity.6A_8"           "Intensity.6A_9"          </span></span>
<span><span class="co">##  [59] "Intensity.6B_7"           "Intensity.6B_8"          </span></span>
<span><span class="co">##  [61] "Intensity.6B_9"           "Reverse"                 </span></span>
<span><span class="co">##  [63] "Potential.contaminant"    "id"                      </span></span>
<span><span class="co">##  [65] "Protein.group.IDs"        "Mod..peptide.IDs"        </span></span>
<span><span class="co">##  [67] "Evidence.IDs"             "MS.MS.IDs"               </span></span>
<span><span class="co">##  [69] "Best.MS.MS"               "Oxidation..M..site.IDs"  </span></span>
<span><span class="co">##  [71] "MS.MS.Count"</span></span>
<span><span class="op">(</span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Intensity\\."</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 56 57 58 59 60 61</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We now read these data using the <code>readSummarizedExperiment</code> function. This peptide-level expression data will be imported into R as an instance of class <code>SummarizedExperiment</code>. We also use the <code>fnames</code> argument to set the row-names of the <code>peptides</code> assay to the peptide sequences and specify that the file is a tab-separated table.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readSummarizedExperiment</a></span><span class="op">(</span><span class="va">f</span>, ecol <span class="op">=</span> <span class="va">i</span>,</span>
<span>                                     fnames <span class="op">=</span> <span class="st">"Sequence"</span>,</span>
<span>                                     sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">cptac_se</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 11466 6 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(11466): AAAAGAGGAGDSGDAVTK AAAALAGGK ... YYTVFDRDNNR</span></span>
<span><span class="co">##    YYTVFDRDNNRVGFAEAAR</span></span>
<span><span class="co">##  rowData names(65): Sequence N.term.cleavage.window ...</span></span>
<span><span class="co">##    Oxidation..M..site.IDs MS.MS.Count</span></span>
<span><span class="co">##  colnames(6): Intensity.6A_7 Intensity.6A_8 ... Intensity.6B_8</span></span>
<span><span class="co">##    Intensity.6B_9</span></span>
<span><span class="co">##  colData names(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Before proceeding, we are going to clean up the sample names by removing the unnecessary <em>Intensity</em> prefix and annotate the experiment in the object’s <code>colData</code>.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"I.+\\."</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cptac_se</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"_[7-9]"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cptac_se</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"^.+_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">##         condition          id</span></span>
<span><span class="co">##       &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##  6A_7          6A           7</span></span>
<span><span class="co">##  6A_8          6A           8</span></span>
<span><span class="co">##  6A_9          6A           9</span></span>
<span><span class="co">##  6B_7          6B           7</span></span>
<span><span class="co">##  6B_8          6B           8</span></span>
<span><span class="co">##  6B_9          6B           9</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>There are many row variables that aren’t useful here. Get rid or all of them but <code>Sequence</code>, <code>Proteins</code>, <code>Leading.razor.protein</code>, <code>PEP</code>, <code>Score</code>, <code>Reverse</code>, and <code>Potential.contaminant</code>.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">keep_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sequence"</span>, <span class="st">"Proteins"</span>, <span class="st">"Leading.razor.protein"</span>, <span class="st">"PEP"</span>,</span>
<span>              <span class="st">"Score"</span>, <span class="st">"Reverse"</span>, <span class="st">"Potential.contaminant"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">[</span>, <span class="va">keep_var</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="analysis-pipeline" class="level2" data-number="4.4"><h2 data-number="4.4" class="anchored" data-anchor-id="analysis-pipeline">
<span class="header-section-number">4.4</span> Analysis pipeline</h2>
<p>A typical quantitative proteomics data processing is composed of the following steps, which we are going to apply to the cptac data created above.</p>
<ul>
<li>Data import</li>
<li>Exploratory data analysis (PCA)</li>
<li>Missing data management (filtering and/or imputation)</li>
<li>Data cleaning</li>
<li>Transformation and normalisation</li>
<li>Aggregation</li>
<li>Downstream analysis</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="missing-values" class="level3" data-number="4.4.1"><h3 data-number="4.4.1" class="anchored" data-anchor-id="missing-values">
<span class="header-section-number">4.4.1</span> Missing values</h3>
<p>Missing values can be highly frequent in proteomics. There are two reasons supporting the existence of missing values, namely biological or technical.</p>
<ol type="1">
<li><p>Values that are missing due to the absence (or extremely low concentration) of a protein are observed for biological reasons, and their pattern <strong>aren’t random</strong> (MNAR). A protein missing due to the suppression of its expression will not be missing at random: it will be missing in the condition in which it was suppressed, and be present in the condition where it is expressed.</p></li>
<li><p>Due to its data-dependent acquisition, mass spectrometry isn’t capable of assaying all peptides in a sample. Peptides that are less abundant than some of their co-eluting ions, peptides that do not ionise well or peptides that do not get identified might be sporadically missing in the final quantitation table, despite their presence in the biological samples. Their absence patterns are (completely) <strong>random</strong> (MAR or MCAR) in such cases.</p></li>
</ol>
<p>Often, third party software that produce quantitative data use zeros instead of properly reporting missing values. We can use the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA()</a></code> function to replace the <code>0</code> by <code>NA</code> values in our <code>cptac_se</code> object and then explore the missing data patterns across columns and rows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span></span>
<span><span class="co">##  $nNA</span></span>
<span><span class="co">##  DataFrame with 1 row and 2 columns</span></span>
<span><span class="co">##          nNA       pNA</span></span>
<span><span class="co">##    &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  1     31130  0.452497</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $nNArows</span></span>
<span><span class="co">##  DataFrame with 11466 rows and 3 columns</span></span>
<span><span class="co">##                 name       nNA       pNA</span></span>
<span><span class="co">##          &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  1     AAAAGAGGAG...         4  0.666667</span></span>
<span><span class="co">##  2         AAAALAGGK         0  0.000000</span></span>
<span><span class="co">##  3        AAAALAGGKK         0  0.000000</span></span>
<span><span class="co">##  4     AAADALSDLE...         0  0.000000</span></span>
<span><span class="co">##  5     AAADALSDLE...         0  0.000000</span></span>
<span><span class="co">##  ...             ...       ...       ...</span></span>
<span><span class="co">##  11462 YYSIYDLGNN...         6  1.000000</span></span>
<span><span class="co">##  11463 YYTFNGPNYN...         3  0.500000</span></span>
<span><span class="co">##  11464    YYTITEVATR         4  0.666667</span></span>
<span><span class="co">##  11465 YYTVFDRDNN...         6  1.000000</span></span>
<span><span class="co">##  11466 YYTVFDRDNN...         6  1.000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $nNAcols</span></span>
<span><span class="co">##  DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##           name       nNA       pNA</span></span>
<span><span class="co">##    &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  1        6A_7      4743  0.413658</span></span>
<span><span class="co">##  2        6A_8      5483  0.478196</span></span>
<span><span class="co">##  3        6A_9      5320  0.463980</span></span>
<span><span class="co">##  4        6B_7      4721  0.411739</span></span>
<span><span class="co">##  5        6B_8      5563  0.485174</span></span>
<span><span class="co">##  6        6B_9      5300  0.462236</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/imagena-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Distribution of missing value (white). Peptides row with more missing values are moved towards the top of the figure.</figcaption></figure>
</div>
</div>
</div>
<p>Let’s now explore these missing values:</p>
<ul>
<li>Explore the number or proportion of missing values across peptides and samples of the <code>cptac_se</code> data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span><span class="op">$</span><span class="va">pNA</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/na2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">nNArows</span><span class="op">$</span><span class="va">nNA</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##     0    1    2    3    4    5    6 </span></span>
<span><span class="co">##  4059  990  884  717  934  807 3075</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Remove rows that have <em>too many</em> missing values. You can do this by hand or using the <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterNA()</a></code> function.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## remove rows that have 4 or more NAs out of 6</span></span>
<span><span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterNA</a></span><span class="op">(</span><span class="va">cptac_se</span>, pNA <span class="op">=</span> <span class="fl">4</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="imputation" class="level3" data-number="4.4.2"><h3 data-number="4.4.2" class="anchored" data-anchor-id="imputation">
<span class="header-section-number">4.4.2</span> Imputation</h3>
<p>Imputation is the technique of replacing missing data with probable values. This can be done with <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute()</a></code> method. As we have discussed above, there are however two types of missing values in mass spectrometry-based proteomics, namely data missing at random (MAR), and data missing not at random (MNAR). These two types of missing data, those missing at random, and those missing not at random, need to be imputed with <a href="https://rformassspectrometry.github.io/QFeatures/articles/Processing.html#imputation-1">different types of imputation methods</a> <span class="citation" data-cites="Lazar:2016">(<a href="#ref-Lazar:2016" role="doc-biblioref">Lazar et al. 2016</a>)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/miximp-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation.</figcaption></figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/imp-sim.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Effect of the nature of missing values on their imputation. Root-mean-square error (RMSE) observations standard deviation ratio (RSR), KNN and MinDet imputation. Lower (blue) is better.</figcaption></figure>
</div>
</div>
</div>
<p>Generally, it is recommended to use <strong>hot deck</strong> methods (nearest neighbour (<strong>left</strong>), maximum likelihood, …) when data are missing at random.Conversely, MNAR features should ideally be imputed with a <strong>left-censor</strong> (minimum value (<strong>right</strong>), but not zero, …) method.</p>
<p>There are various methods to perform data imputation, as described in <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">?impute</a></code>. The <em><a href="https://CRAN.R-project.org/package=imp4p">imp4p</a></em> package contains additional functionality, including some to estimate the randomness of missing data.</p>
<p>The general syntax for imputation is shown below, using the <code>se_na2</code> object as an example:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span></span>
<span><span class="co">## impute missing values using knn imputation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span><span class="co">##  Loading required namespace: impute</span></span>
<span><span class="co">##  Imputing along margin 1 (features/rows).</span></span>
<span><span class="co">##  Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##   mean imputation used for these rows</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 689 16 </span></span>
<span><span class="co">##  metadata(3): MSnbaseFiles MSnbaseProcessing MSnbaseVersion</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(689): AT1G09210 AT1G21750 ... AT4G11150 AT4G39080</span></span>
<span><span class="co">##  rowData names(2): nNA randna</span></span>
<span><span class="co">##  colnames(16): M1F1A M1F4A ... M2F8B M2F11B</span></span>
<span><span class="co">##  colData names(1): nNA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Following the example above, apply a mixed imputation, using knn for data missing at random and the zero imputation for data missing not at random. Hint: the <code>randna</code> variable defines which features are assumed to be missing at random.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, <span class="st">"mixed"</span>,</span>
<span>       randna <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">$</span><span class="va">randna</span>,</span>
<span>       mar <span class="op">=</span> <span class="st">"knn"</span>, mnar <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span></span>
<span><span class="co">##  class: SummarizedExperiment </span></span>
<span><span class="co">##  dim: 689 16 </span></span>
<span><span class="co">##  metadata(3): MSnbaseFiles MSnbaseProcessing MSnbaseVersion</span></span>
<span><span class="co">##  assays(1): ''</span></span>
<span><span class="co">##  rownames(689): AT1G09210 AT1G21750 ... AT4G11150 AT4G39080</span></span>
<span><span class="co">##  rowData names(2): nNA randna</span></span>
<span><span class="co">##  colnames(16): M1F1A M1F4A ... M2F8B M2F11B</span></span>
<span><span class="co">##  colData names(1): nNA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>When assessing missing data imputation methods, such as in <a href="https://pubs.acs.org/doi/abs/10.1021/acs.jproteome.5b00981">Lazar et al.&nbsp;(2016)</a>, one often replaces values with missing data, imputes these with a method of choice, then quantifies the difference between original (expected) and observed (imputed) values. Here, using the <code>se_na2</code> data, use this strategy to assess the difference between knn and Bayesian PCA imputation.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="30-quant_cache/html/naex2_1f97ee4eda55cca99ebc3c0528ff05d3">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">imp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span><span class="co">##  Imputing along margin 1 (features/rows).</span></span>
<span><span class="co">##  Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##   mean imputation used for these rows</span></span>
<span><span class="va">imp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"bpca"</span><span class="op">)</span></span>
<span><span class="co">##  Loading required namespace: pcaMethods</span></span>
<span><span class="co">##  Imputing along margin 1 (features/rows).</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">imp1</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">imp2</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##       Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">##  5.332e-05 6.594e-03 1.535e-02 2.315e-02 2.855e-02 2.579e-01</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html">na.omit</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##     Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##   0.0170  0.1865  0.2440  0.2500  0.3080  0.6587</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>When assessing the impact of missing value imputation on real data, one can’t use the strategy above. Another useful approach is to assess the impact of the imputation method on the distribution of the quantitative data. For instance, here is the intensity distribution of the <code>se_na2</code> data. Verify the effect of applying <code>knn</code>, <code>zero</code>, <code>MinDet</code> and <code>bpca</code> on this distribution.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html">na.omit</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/nasetdist-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Intensity disctribution of the naset data.</figcaption></figure>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="30-quant_cache/html/naex3_895f03476cf7aea332cbb7057c138bd6">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"steelblue"</span>, <span class="st">"orange"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html">na.omit</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se_na2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">cls</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">cls</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  Imputing along margin 1 (features/rows).</span></span>
<span><span class="co">##  Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##   mean imputation used for these rows</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">cls</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"MinDet"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">cls</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  Imputing along margin 2 (samples/columns).</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">impute</a></span><span class="op">(</span><span class="va">se_na2</span>, method <span class="op">=</span> <span class="st">"bpca"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">cls</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##  Imputing along margin 1 (features/rows).</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orig"</span>, <span class="st">"knn"</span>, <span class="st">"zero"</span>, <span class="st">"MinDet"</span>, <span class="st">"bpca"</span><span class="op">)</span>,</span>
<span>       col <span class="op">=</span> <span class="va">cls</span>, lwd <span class="op">=</span> <span class="fl">2</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/naex3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p><strong>Tip</strong>: When downstream analyses permit, it might be safer not to impute data and deal explicitly with missing values. Indeed missing data imputation is not straightforward, and is likely to dramatically fail when a high proportion of data is missing (10s of %). It is possible to keep NAs when performing hypothesis tests<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, but (generally) not to perform a principal component analysis.</p>
</section><section id="identification-quality-control" class="level3" data-number="4.4.3"><h3 data-number="4.4.3" class="anchored" data-anchor-id="identification-quality-control">
<span class="header-section-number">4.4.3</span> Identification quality control</h3>
<p>As discussed in the previous chapter, PSMs are deemed relevant after comparison against hits from a decoy database. The origin of these hits is recorded with <code>+</code> in the <code>Reverse</code> variable:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##          + </span></span>
<span><span class="co">##  7572   12</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Similarly, a proteomics experiment is also searched against a database of contaminants:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">Potential.contaminant</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##          + </span></span>
<span><span class="co">##  7558   26</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s visualise some of the cptac’s metadata using standard <code>ggplot2</code> code:</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Visualise the identification score and the posterior probability probability (PEP) distributions from forward and reverse hits and interpret the figure.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Score</span>, colour <span class="op">=</span> <span class="va">Reverse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/idqc1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PEP</span>, colour <span class="op">=</span> <span class="va">Reverse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/idqc2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p><strong>Note</strong>: it is also possible to compute and visualise protein groups as connected components starting from a quantitative dataset such as a <code>SummarizedExperiment</code>. See the <a href="https://rformassspectrometry.github.io/PSMatch/articles/AdjacencyMatrix.html#using-quantitative-data"><em>Using quantitative data</em></a> section in the <em>Understanding protein groups with adjacency matrices</em> vignette.</p>
</section><section id="creating-the-qfeatures-data" class="level3" data-number="4.4.4"><h3 data-number="4.4.4" class="anchored" data-anchor-id="creating-the-qfeatures-data">
<span class="header-section-number">4.4.4</span> Creating the QFeatures data</h3>
<p>We can now create our <code>QFeatures</code> object using the <code>SummarizedExperiment</code> as shown below.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">QFeatures</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>peptides <span class="op">=</span> <span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cptac</span></span>
<span><span class="co">##  An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">##   [1] peptides: SummarizedExperiment with 7584 rows and 6 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We should also assign the <code>QFeatures</code> column data with the <code>SummarizedExperiment</code> slot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">cptac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that it is also possible to directly create a <code>QFeatures</code> object with the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readQFeatures()</a></code> function and the same arguments as the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readSummarizedExperiment()</a></code> used above. In addition, most functions used above and below work on single <code>SummarizedExperiment</code> objects or assays within a <code>QFeatures</code> object.</p>
</section><section id="filtering-out-contaminants-and-reverse-hits" class="level3" data-number="4.4.5"><h3 data-number="4.4.5" class="anchored" data-anchor-id="filtering-out-contaminants-and-reverse-hits">
<span class="header-section-number">4.4.5</span> Filtering out contaminants and reverse hits</h3>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures()</a></code> function, filter out the reverse and contaminant hits, and also retain those that have a posterior error probability smaller than 0.05.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">cptac</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">PEP</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">##  'Reverse' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">##  'Potential.contaminant' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">##  'PEP' found in 1 out of 1 assay(s)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="log-transformation-and-normalisation" class="level3" data-number="4.4.6"><h3 data-number="4.4.6" class="anchored" data-anchor-id="log-transformation-and-normalisation">
<span class="header-section-number">4.4.6</span> Log-transformation and normalisation</h3>
<p>The two code chunks below log-transform and normalise using the assay <code>i</code> as input and adding a new one names as defined by <code>name</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform</a></span><span class="op">(</span><span class="va">cptac</span>, i <span class="op">=</span> <span class="st">"peptides"</span>,</span>
<span>                      name <span class="op">=</span> <span class="st">"log_peptides"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html">normalize()</a></code> method to normalise the data. The syntax is the same as <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform()</a></code>. Use the <code>center.median</code> method.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html">normalize</a></span><span class="op">(</span><span class="va">cptac</span>, i <span class="op">=</span> <span class="st">"log_peptides"</span>,</span>
<span>                   name <span class="op">=</span> <span class="st">"lognorm_peptides"</span>,</span>
<span>                   method <span class="op">=</span> <span class="st">"center.median"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Visualise the result of the transformations above. The <code><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities()</a></code> function from the <code>limma</code> package is very convenient, but feel free to use boxplots, violin plots, or any other visualisation that you deem useful to assess the tranformations.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"log_peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"lognorm_peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/plotdens-1.png" class="img-fluid figure-img" width="1440"></p>
<figcaption class="figure-caption">Three peptide level assays: raw data, log transformed and normalised.</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="aggregation" class="level3" data-number="4.4.7"><h3 data-number="4.4.7" class="anchored" data-anchor-id="aggregation">
<span class="header-section-number">4.4.7</span> Aggregation</h3>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use median aggregation to aggregation peptides into protein values. This is not necessarily the best choice, as we will see later, but a good start.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-29-contents" aria-controls="callout-29" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-29" class="callout-29-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cptac</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">aggregateFeatures</a></span><span class="op">(</span><span class="va">cptac</span>,</span>
<span>                      <span class="st">"lognorm_peptides"</span>,</span>
<span>                      name <span class="op">=</span> <span class="st">"proteins_med"</span>,</span>
<span>                      fcol <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span>                      fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Looking at the <code>.n</code> row variable computed during the aggregation, we see that most proteins result from the aggregation of 5 peptides or less, while very few proteins are accounted for by tens of peptides.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"proteins_med"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">.n</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##    1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19 </span></span>
<span><span class="co">##  327 234 167 132  84  73  62  49  49  29  29  24  20  13  15  12   4   6  11 </span></span>
<span><span class="co">##   20  21  22  23  24  25  26  28  29  30  31  32  34  37  38  39  42  51  52 </span></span>
<span><span class="co">##    5   7   4   7   2   2   3   1   3   1   2   2   1   1   1   1   2   1   1 </span></span>
<span><span class="co">##   62 </span></span>
<span><span class="co">##    1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="principal-component-analysis" class="level3" data-number="4.4.8"><h3 data-number="4.4.8" class="anchored" data-anchor-id="principal-component-analysis">
<span class="header-section-number">4.4.8</span> Principal component analysis</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sthda.com/english/rpkgs/factoextra">factoextra</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_pep</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">cptac</span><span class="op">[[</span><span class="st">"lognorm_peptides"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterNA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">assay</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html">fviz_pca_ind</a></span><span class="op">(</span>habillage <span class="op">=</span> <span class="va">cptac</span><span class="op">$</span><span class="va">condition</span>, title <span class="op">=</span> <span class="st">"Peptides"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_prot</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">cptac</span><span class="op">[[</span><span class="st">"proteins_med"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">filterNA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">assay</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html">fviz_pca_ind</a></span><span class="op">(</span>habillage <span class="op">=</span> <span class="va">cptac</span><span class="op">$</span><span class="va">condition</span>,</span>
<span>                 title <span class="op">=</span> <span class="st">"Proteins (median aggregation)"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">pca_pep</span> <span class="op">+</span> <span class="va">pca_prot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/plotpca-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Peptide and protein level PCA analyses.</figcaption></figure>
</div>
</div>
</div>
</section><section id="visualisation" class="level3" data-number="4.4.9"><h3 data-number="4.4.9" class="anchored" data-anchor-id="visualisation">
<span class="header-section-number">4.4.9</span> Visualisation</h3>
<p>Below, we use the <code>longFormat()</code> function to extract the quantitative and row data in a long format, that can be directly reused by the tidyverse tools.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">longFormat</span><span class="op">(</span><span class="va">cptac</span><span class="op">[</span><span class="st">"P02787ups|TRFE_HUMAN_UPS"</span>, ,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lognorm_peptides"</span>, <span class="st">"proteins_med"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="va">colname</span><span class="op">)</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colname</span>, y <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">rowname</span>, shape <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="op">~</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"P02787ups|TRFE_HUMAN_UPS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/vis-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Peptide and protein expression profile.</figcaption></figure>
</div>
</div>
</div>
<p>We can also visualise the assays withing a <code>QFeatures</code> object and their relation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">cptac</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/plotqf-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>The example above shows a simple linear relationship between assays. Create a more interesting one by applying a different normalisation method on the <em>log_peptides</em> assay and aggreate that new normalised peptide assay. Visualise the relationship with <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code>, as above.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-31-contents" aria-controls="callout-31" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-31" class="callout-31-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html">normalize</a></span><span class="op">(</span><span class="va">cptac</span>, <span class="st">"log_peptides"</span>,</span>
<span>          name <span class="op">=</span> <span class="st">"logquantiles_peptides"</span>,</span>
<span>          method <span class="op">=</span> <span class="st">"quantiles"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">aggregateFeatures</a></span><span class="op">(</span></span>
<span>        <span class="st">"logquantiles_peptides"</span>,</span>
<span>        name <span class="op">=</span> <span class="st">"proteins_med2"</span>,</span>
<span>        fcol <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span>        fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>        na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30-quant_files/figure-html/plotqf2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="statistical-analysis" class="level3" data-number="4.4.10"><h3 data-number="4.4.10" class="anchored" data-anchor-id="statistical-analysis">
<span class="header-section-number">4.4.10</span> Statistical analysis</h3>
<p>R in general and Bioconductor in particular are well suited for the statistical analysis of quantitative proteomics data. Several packages provide dedicated resources for proteomics data:</p>
<ul>
<li><p><em><a href="https://bioconductor.org/packages/3.19/MSstats">MSstats</a></em> and <em><a href="https://bioconductor.org/packages/3.19/MSstatsTMT">MSstatsTMT</a></em>: A set of tools for statistical relative protein significance analysis in Data dependent (DDA), SRM, Data independent acquisition (DIA) and TMT experiments.</p></li>
<li><p><em><a href="https://bioconductor.org/packages/3.19/msmsTests">msmsTests</a></em>: Statistical tests for label-free LC-MS/MS data by spectral counts, to discover differentially expressed proteins between two biological conditions. Three tests are available: Poisson GLM regression, quasi-likelihood GLM regression, and the negative binomial of the <em><a href="https://bioconductor.org/packages/3.19/edgeR">edgeR</a></em> package. All can be readily applied on <code>MSnSet</code> instances produced, for example by <code>MSnID</code>.</p></li>
<li><p><em><a href="https://bioconductor.org/packages/3.19/DEP">DEP</a></em> provides an integrated analysis workflow for the analysis of mass spectrometry proteomics data for differential protein expression or differential enrichment.</p></li>
<li><p><em><a href="https://github.com/statOmics/MSqRob">MSqRob</a></em>: The <code>MSqRob</code> package allows a user to do quantitative protein-level statistical inference on LC-MS proteomics data. More specifically, our package makes use of peptide-level input data, thus correcting for unbalancedness and peptide-specific biases. As previously shown (<a href="https://pubs.acs.org/doi/abs/10.1021/pr501223t">Goeminne et al.&nbsp;(2015)</a>), this approach is both more sensitive and specific than summarizing peptide-level input to protein-level values. Model estimates are stabilized by ridge regression, empirical Bayes variance estimation and downweighing of outliers. Currently, only label-free proteomics data types are supported. <a href="https://github.com/statOmics/msqrob2/"><code>msqrob2</code></a> is now available and makes use of the <code>QFeatures</code> infrastructure.</p></li>
<li><p><em><a href="https://github.com/const-ae/proDA">proDA</a></em> accounts for missing values in label-free mass spectrometry data without imputation. The package implements a probabilistic dropout model that ensures that the information from observed and missing values are properly combined. It adds empirical Bayesian priors to increase power to detect differentially abundant proteins.</p></li>
</ul>
<p>Others, while not specfic to proteomics, are also recommended, such as the <em><a href="https://bioconductor.org/packages/3.19/limma">limma</a></em> package. When analysing spectral counting data, methods for high throughput sequencing data are applicable. Below, we illustrate how to apply a typical <code>edgeR</code> test to count data using the <code>msms.edgeR</code> function from the <code>msmsTests</code> package.</p>
<p>Below, we are going to perform our statistical analysis on the protein data using <code>limma</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html">getWithColData</a></span><span class="op">(</span><span class="va">cptac</span>, <span class="st">"proteins_med"</span><span class="op">)</span></span>
<span><span class="co">##  Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">##  Warning: Ignoring redundant column names in 'colData(x)':</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <em><a href="https://bioconductor.org/packages/3.19/limma">limma</a></em> package is the precursor package that enables the consistent application of linear models to normalliy distributed omics data in general, and microarrays in particular.</p>
<p>The <code>limma</code> package implements an empirical Bayes method that borrows information across features to estimate the standard error and calculate (so called moderated) t statistics. This approach is demonstrably more powerful that a standard t-tests when the number of samples is low.</p>
<p>The code chunk below illustrates how to set up the model, fit it, and apply the empirical Bayes moderation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">prots</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">prots</span><span class="op">)</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="co">##  Warning: Partial NA coefficients for 25 probe(s)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, the <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable()</a></code> function is used the extract the results for the coefficient of interest.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"prots$condition6B"</span>, number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"protein"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>TP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"ups"</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note the warning about partial <code>NA</code> coefficients for 23 probes:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">na_coefs</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">protein</span><span class="op">)</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">prots</span><span class="op">[</span><span class="va">na_coefs</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##                                 6A_7      6A_8       6A_9       6B_7</span></span>
<span><span class="co">##  P00167ups|CYB5_HUMAN_UPS        NaN       NaN        NaN -0.7840558</span></span>
<span><span class="co">##  P01112ups|RASH_HUMAN_UPS        NaN       NaN        NaN -1.5564896</span></span>
<span><span class="co">##  P05413ups|FABPH_HUMAN_UPS       NaN       NaN        NaN -3.3419480</span></span>
<span><span class="co">##  P08758ups|ANXA5_HUMAN_UPS       NaN       NaN        NaN -2.7973872</span></span>
<span><span class="co">##  sp|P06704|CDC31_YEAST           NaN       NaN        NaN -1.2032046</span></span>
<span><span class="co">##  sp|P25574|EMC1_YEAST      -1.506177 -1.983737 -0.7795009        NaN</span></span>
<span><span class="co">##  sp|P32608|RTG2_YEAST            NaN       NaN        NaN        NaN</span></span>
<span><span class="co">##  sp|P32769|HBS1_YEAST            NaN -1.384031 -0.7285780        NaN</span></span>
<span><span class="co">##  sp|P34217|PIN4_YEAST            NaN       NaN        NaN -0.8378614</span></span>
<span><span class="co">##  sp|P34237|CASP_YEAST            NaN       NaN        NaN -1.5645172</span></span>
<span><span class="co">##  sp|P38166|SFT2_YEAST      -1.585685 -1.076707        NaN        NaN</span></span>
<span><span class="co">##  sp|P40056|GET2_YEAST            NaN -1.091696 -1.4014211        NaN</span></span>
<span><span class="co">##  sp|P40533|TED1_YEAST            NaN       NaN        NaN -2.0491876</span></span>
<span><span class="co">##  sp|P43582|WWM1_YEAST            NaN       NaN        NaN -0.5538711</span></span>
<span><span class="co">##  sp|P46965|SPC1_YEAST            NaN -3.428771 -3.6321984        NaN</span></span>
<span><span class="co">##  sp|P48363|PFD3_YEAST            NaN       NaN        NaN -0.1904905</span></span>
<span><span class="co">##  sp|P53044|UFD1_YEAST            NaN       NaN        NaN -2.3937896</span></span>
<span><span class="co">##  sp|P53091|MCM6_YEAST            NaN -1.253779 -1.6910302        NaN</span></span>
<span><span class="co">##  sp|P53115|INO80_YEAST           NaN       NaN        NaN -2.7183960</span></span>
<span><span class="co">##  sp|P53893|RIA1_YEAST      -1.830856       NaN -1.5232812        NaN</span></span>
<span><span class="co">##  sp|Q05029|BCH1_YEAST      -1.857855 -1.760501        NaN        NaN</span></span>
<span><span class="co">##  sp|Q07540|FRDA_YEAST            NaN       NaN        NaN -2.3850242</span></span>
<span><span class="co">##  sp|Q08645|FOLE_YEAST            NaN       NaN        NaN -1.7269184</span></span>
<span><span class="co">##  sp|Q12255|NYV1_YEAST            NaN       NaN        NaN -2.8706030</span></span>
<span><span class="co">##  sp|Q12280|IQG1_YEAST            NaN       NaN        NaN -2.5636389</span></span>
<span><span class="co">##                                  6B_8       6B_9</span></span>
<span><span class="co">##  P00167ups|CYB5_HUMAN_UPS  -2.0282987 -1.1230809</span></span>
<span><span class="co">##  P01112ups|RASH_HUMAN_UPS         NaN -1.5618192</span></span>
<span><span class="co">##  P05413ups|FABPH_HUMAN_UPS        NaN -3.8907081</span></span>
<span><span class="co">##  P08758ups|ANXA5_HUMAN_UPS -2.0137585 -2.0894752</span></span>
<span><span class="co">##  sp|P06704|CDC31_YEAST     -2.1252371 -1.5844104</span></span>
<span><span class="co">##  sp|P25574|EMC1_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P32608|RTG2_YEAST      -4.4424189 -2.7873186</span></span>
<span><span class="co">##  sp|P32769|HBS1_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P34217|PIN4_YEAST      -0.1316397 -0.1989392</span></span>
<span><span class="co">##  sp|P34237|CASP_YEAST      -1.6600291 -1.6877463</span></span>
<span><span class="co">##  sp|P38166|SFT2_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P40056|GET2_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P40533|TED1_YEAST             NaN -1.7474812</span></span>
<span><span class="co">##  sp|P43582|WWM1_YEAST      -0.7360990 -0.7207043</span></span>
<span><span class="co">##  sp|P46965|SPC1_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P48363|PFD3_YEAST             NaN -0.5087747</span></span>
<span><span class="co">##  sp|P53044|UFD1_YEAST      -1.5138746        NaN</span></span>
<span><span class="co">##  sp|P53091|MCM6_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|P53115|INO80_YEAST            NaN -2.7504079</span></span>
<span><span class="co">##  sp|P53893|RIA1_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|Q05029|BCH1_YEAST             NaN        NaN</span></span>
<span><span class="co">##  sp|Q07540|FRDA_YEAST             NaN -2.0325180</span></span>
<span><span class="co">##  sp|Q08645|FOLE_YEAST             NaN -1.5257358</span></span>
<span><span class="co">##  sp|Q12255|NYV1_YEAST             NaN -2.8237317</span></span>
<span><span class="co">##  sp|Q12280|IQG1_YEAST      -2.0547564 -2.4648042</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now visualise the results using a volcano plot:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">TP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using the pipeline described above, we would would identify a single differentially expressed protein at an 5 percent FDR but miss out the other 36 expected spike-in proteins.</p>
<p>We can assess our results in terms of true/false postitves/negatives:</p>
<ul>
<li>True positives: 1</li>
<li>False positives: 0</li>
<li>True negatives: 1330</li>
<li>False negatives: 32</li>
</ul></section></section><section id="summary-exercice" class="level2" data-number="4.5"><h2 data-number="4.5" class="anchored" data-anchor-id="summary-exercice">
<span class="header-section-number">4.5</span> Summary exercice</h2>
<p>As shown below, it is possible to substantially improve these results by aggregating features using a robust summarisation (available as <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html">MsCoreUtils::robustSummary()</a></code>), i.e robust regression with M-estimation using Huber weights, as described in section 2.7 in <span class="citation" data-cites="Sticker:2019">(<a href="#ref-Sticker:2019" role="doc-biblioref">Sticker et al. 2019</a>)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="./img/vp2.png" class="img-fluid figure-img" width="1048"></p>
<figcaption class="figure-caption">Aggregation using robust summarisation.</figcaption></figure>
</div>
</div>
</div>
<ul>
<li>True positives: 21</li>
<li>False positives: 2</li>
<li>True negatives: 1340</li>
<li>False negatives: 12</li>
</ul>
<p>Repeat and adapt what we have seen here using, for example, the <code>robustSummary()</code> function.</p>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Christoforou:2016" class="csl-entry" role="listitem">
Christoforou, Andy, Claire M Mulvey, Lisa M Breckels, Aikaterini Geladaki, Tracey Hurrell, Penelope C Hayward, Thomas Naake, et al. 2016. <span>“A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.”</span> <em>Nat Commun</em> 7: 8992. <a href="https://doi.org/10.1038/ncomms9992">https://doi.org/10.1038/ncomms9992</a>.
</div>
<div id="ref-Cox:2008" class="csl-entry" role="listitem">
Cox, J, and M Mann. 2008. <span>“MaxQuant Enables High Peptide Identification Rates, Individualized p.p.b.-Range Mass Accuracies and Proteome-Wide Protein Quantification.”</span> <em>Nat Biotechnol</em> 26 (12): 1367–72. <a href="https://doi.org/10.1038/nbt.1511">https://doi.org/10.1038/nbt.1511</a>.
</div>
<div id="ref-Lazar:2016" class="csl-entry" role="listitem">
Lazar, C, L Gatto, M Ferro, C Bruley, and T Burger. 2016. <span>“Accounting for the Multiple Natures of Missing Values in Label-Free Quantitative Proteomics Data Sets to Compare Imputation Strategies.”</span> <em>J Proteome Res</em> 15 (4): 1116–25. <a href="https://doi.org/10.1021/acs.jproteome.5b00981">https://doi.org/10.1021/acs.jproteome.5b00981</a>.
</div>
<div id="ref-SE" class="csl-entry" role="listitem">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment">https://bioconductor.org/packages/SummarizedExperiment</a>.
</div>
<div id="ref-Paulovich:2010" class="csl-entry" role="listitem">
Paulovich, Amanda G, Dean Billheimer, Amy-Joan L Ham, Lorenzo Vega-Montoto, Paul A Rudnick, David L Tabb, Pei Wang, et al. 2010. <span>“Interlaboratory Study Characterizing a Yeast Performance Standard for Benchmarking <span>LC-MS</span> Platform Performance.”</span> <em>Mol. Cell. Proteomics</em> 9 (2): 242–54.
</div>
<div id="ref-MAE" class="csl-entry" role="listitem">
Ramos, Marcel, Lucas Schiffer, Angela Re, Rimsha Azhar, Azfar Basunia, Carmen Rodriguez Cabrera, Tiffany Chan, et al. 2017. <span>“Software for the Integration of Multi-Omics Experiments in Bioconductor.”</span> <em>Cancer Research</em> 77(21); e39-42.
</div>
<div id="ref-Sticker:2019" class="csl-entry" role="listitem">
Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement. 2019. <span>“Robust Summarization and Inference in Proteome-Wide Label-Free Quantification.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/668863">https://doi.org/10.1101/668863</a>.
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>Still, it is recommended to explore missingness as part of the exploratory data analysis.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/20-id.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/95-annex.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="../LICENSE.html">License</a>
  </li>  
</ul>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>