<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R for Mass Spectrometry - 5&nbsp; Annex</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/99-si.html" rel="next">
<link href="../pages/30-quant.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/95-annex.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">R for Mass Spectrometry</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/R4MS/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/R4MS/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/raw-ms-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Raw MS data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20-id.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identification data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30-quant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95-annex.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/99-si.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Additional materials and help</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-raw2" id="toc-sec-raw2" class="nav-link active" data-scroll-target="#sec-raw2"><span class="header-section-number">5.1</span> Raw MS data under the hood: the <code>mzR</code> package</a></li>
  <li>
<a href="#sec-id2" id="toc-sec-id2" class="nav-link" data-scroll-target="#sec-id2"><span class="header-section-number">5.2</span> PSM data under the hood</a>
  <ul class="collapse">
<li><a href="#mzid" id="toc-mzid" class="nav-link" data-scroll-target="#mzid"><code>mzID</code></a></li>
  <li><a href="#mzr" id="toc-mzr" class="nav-link" data-scroll-target="#mzr"><code>mzR</code></a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-anx" class="quarto-section-identifier"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Annex</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><section id="sec-raw2" class="level2" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="sec-raw2">
<span class="header-section-number">5.1</span> Raw MS data under the hood: the <code>mzR</code> package</h2>
<p>The <code>mzR</code> package is a direct interface to the <a href="http://proteowizard.sourceforge.net/">proteowizard</a> code base. It includes a substantial proportion of <em>pwiz</em>’s C/C++ code for fast and efficient parsing of these large raw data files.</p>
<p>Let’s start by using some raw data files from the <code>msdata</code> package. After loading it, we use the <code>proteomics()</code> function to return the full file names for two raw data files. We will start by focusing on the second one.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">proteomics</a></span><span class="op">(</span>full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">f</span></span>
<span><span class="co">##  [1] "/usr/local/lib/R/site-library/msdata/proteomics/MRM-standmix-5.mzML.gz"                                                </span></span>
<span><span class="co">##  [2] "/usr/local/lib/R/site-library/msdata/proteomics/MS3TMT10_01022016_32917-33481.mzML.gz"                                 </span></span>
<span><span class="co">##  [3] "/usr/local/lib/R/site-library/msdata/proteomics/MS3TMT11.mzML"                                                         </span></span>
<span><span class="co">##  [4] "/usr/local/lib/R/site-library/msdata/proteomics/TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz"</span></span>
<span><span class="co">##  [5] "/usr/local/lib/R/site-library/msdata/proteomics/TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.mzML.gz"</span></span>
<span><span class="op">(</span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"20141210"</span>, <span class="va">f</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "/usr/local/lib/R/site-library/msdata/proteomics/TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The three main functions of <code>mzR</code> are</p>
<ul>
<li>
<code>openMSfile</code> to create a file handle to a raw data file</li>
<li>
<code>header</code> to extract metadata about the spectra contained in the file</li>
<li>
<code>peaks</code> to extract one or multiple spectra of interest.</li>
</ul>
<p>Other functions such as <code>instrumentInfo</code>, or <code>runInfo</code> can be used to gather general information about a run.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/mzR/">mzR</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: Rcpp</span></span>
<span><span class="va">ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mzR/man/openMSfile.html">openMSfile</a></span><span class="op">(</span><span class="va">f2</span><span class="op">)</span></span>
<span><span class="va">ms</span></span>
<span><span class="co">##  Mass Spectrometry file handle.</span></span>
<span><span class="co">##  Filename:  TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz </span></span>
<span><span class="co">##  Number of scans:  7534</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mzR/man/peaks.html">header</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">hd</span><span class="op">)</span></span>
<span><span class="co">##  [1] 7534   31</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hd</span><span class="op">)</span></span>
<span><span class="co">##   [1] "seqNum"                     "acquisitionNum"            </span></span>
<span><span class="co">##   [3] "msLevel"                    "polarity"                  </span></span>
<span><span class="co">##   [5] "peaksCount"                 "totIonCurrent"             </span></span>
<span><span class="co">##   [7] "retentionTime"              "basePeakMZ"                </span></span>
<span><span class="co">##   [9] "basePeakIntensity"          "collisionEnergy"           </span></span>
<span><span class="co">##  [11] "ionisationEnergy"           "lowMZ"                     </span></span>
<span><span class="co">##  [13] "highMZ"                     "precursorScanNum"          </span></span>
<span><span class="co">##  [15] "precursorMZ"                "precursorCharge"           </span></span>
<span><span class="co">##  [17] "precursorIntensity"         "mergedScan"                </span></span>
<span><span class="co">##  [19] "mergedResultScanNum"        "mergedResultStartScanNum"  </span></span>
<span><span class="co">##  [21] "mergedResultEndScanNum"     "injectionTime"             </span></span>
<span><span class="co">##  [23] "filterString"               "spectrumId"                </span></span>
<span><span class="co">##  [25] "centroided"                 "ionMobilityDriftTime"      </span></span>
<span><span class="co">##  [27] "isolationWindowTargetMZ"    "isolationWindowLowerOffset"</span></span>
<span><span class="co">##  [29] "isolationWindowUpperOffset" "scanWindowLowerLimit"      </span></span>
<span><span class="co">##  [31] "scanWindowUpperLimit"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">peaks</a></span><span class="op">(</span><span class="va">ms</span>, <span class="fl">117</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##             mz intensity</span></span>
<span><span class="co">##  [1,] 399.9976         0</span></span>
<span><span class="co">##  [2,] 399.9991         0</span></span>
<span><span class="co">##  [3,] 400.0006         0</span></span>
<span><span class="co">##  [4,] 400.0021         0</span></span>
<span><span class="co">##  [5,] 400.2955         0</span></span>
<span><span class="co">##  [6,] 400.2970         0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">peaks</a></span><span class="op">(</span><span class="va">ms</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  List of 5</span></span>
<span><span class="co">##   $ : num [1:25800, 1:2] 400 400 400 400 400 ...</span></span>
<span><span class="co">##    ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##    .. ..$ : NULL</span></span>
<span><span class="co">##    .. ..$ : chr [1:2] "mz" "intensity"</span></span>
<span><span class="co">##   $ : num [1:25934, 1:2] 400 400 400 400 400 ...</span></span>
<span><span class="co">##    ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##    .. ..$ : NULL</span></span>
<span><span class="co">##    .. ..$ : chr [1:2] "mz" "intensity"</span></span>
<span><span class="co">##   $ : num [1:26148, 1:2] 400 400 400 400 400 ...</span></span>
<span><span class="co">##    ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##    .. ..$ : NULL</span></span>
<span><span class="co">##    .. ..$ : chr [1:2] "mz" "intensity"</span></span>
<span><span class="co">##   $ : num [1:26330, 1:2] 400 400 400 400 400 ...</span></span>
<span><span class="co">##    ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##    .. ..$ : NULL</span></span>
<span><span class="co">##    .. ..$ : chr [1:2] "mz" "intensity"</span></span>
<span><span class="co">##   $ : num [1:26463, 1:2] 400 400 400 400 400 ...</span></span>
<span><span class="co">##    ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##    .. ..$ : NULL</span></span>
<span><span class="co">##    .. ..$ : chr [1:2] "mz" "intensity"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Let’s extract the index of the MS2 spectrum with the highest base peak intensity and plot its spectrum. Is the data centroided or in profile mode?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hd2</span> <span class="op">&lt;-</span> <span class="va">hd</span><span class="op">[</span><span class="va">hd</span><span class="op">$</span><span class="va">msLevel</span> <span class="op">==</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">hd2</span><span class="op">$</span><span class="va">basePeakIntensity</span><span class="op">)</span></span>
<span><span class="va">hd2</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span></span>
<span><span class="co">##       seqNum acquisitionNum msLevel polarity peaksCount totIonCurrent</span></span>
<span><span class="co">##  5404   5404           5404       2        1        275    2283283712</span></span>
<span><span class="co">##       retentionTime basePeakMZ basePeakIntensity collisionEnergy</span></span>
<span><span class="co">##  5404      2751.313   859.5032         354288224              45</span></span>
<span><span class="co">##       ionisationEnergy    lowMZ  highMZ precursorScanNum precursorMZ</span></span>
<span><span class="co">##  5404                0 100.5031 1995.63             5403    859.1722</span></span>
<span><span class="co">##       precursorCharge precursorIntensity mergedScan mergedResultScanNum</span></span>
<span><span class="co">##  5404               3          627820480         NA                  NA</span></span>
<span><span class="co">##       mergedResultStartScanNum mergedResultEndScanNum injectionTime</span></span>
<span><span class="co">##  5404                       NA                     NA    0.03474091</span></span>
<span><span class="co">##                                                   filterString</span></span>
<span><span class="co">##  5404 FTMS + p NSI d Full ms2 859.50@hcd45.00 [100.00-2000.00]</span></span>
<span><span class="co">##                                          spectrumId centroided</span></span>
<span><span class="co">##  5404 controllerType=0 controllerNumber=1 scan=5404       TRUE</span></span>
<span><span class="co">##       ionMobilityDriftTime isolationWindowTargetMZ isolationWindowLowerOffset</span></span>
<span><span class="co">##  5404                   NA                   859.5                          1</span></span>
<span><span class="co">##       isolationWindowUpperOffset scanWindowLowerLimit scanWindowUpperLimit</span></span>
<span><span class="co">##  5404                          1                  100                 2000</span></span>
<span><span class="va">pi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">peaks</a></span><span class="op">(</span><span class="va">ms</span>, <span class="va">hd2</span><span class="op">[</span><span class="va">i</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pi</span>, type <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="95-annex_files/figure-html/ex_raw-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mz</span> <span class="op">&lt;-</span> <span class="va">hd2</span><span class="op">[</span><span class="va">i</span>, <span class="st">"basePeakMZ"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pi</span>, type <span class="op">=</span> <span class="st">"h"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">mz</span> <span class="op">-</span> <span class="fl">0.5</span>, <span class="va">mz</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="95-annex_files/figure-html/ex_raw-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Pick an MS1 spectrum and visually check whether it is centroided or in profile mode.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Zooming into spectrum 300 (an MS1 spectrum).</span></span>
<span><span class="va">j</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span><span class="va">pj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">peaks</a></span><span class="op">(</span><span class="va">ms</span>, <span class="va">j</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pj</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="95-annex_files/figure-html/ex_raw2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mz</span> <span class="op">&lt;-</span> <span class="va">hd</span><span class="op">[</span><span class="va">j</span>, <span class="st">"basePeakMZ"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pj</span>, type <span class="op">=</span> <span class="st">"l"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">mz</span> <span class="op">-</span> <span class="fl">0.5</span>, <span class="va">mz</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="95-annex_files/figure-html/ex_raw2-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="sec-id2" class="level2" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="sec-id2">
<span class="header-section-number">5.2</span> PSM data under the hood</h2>
<p>There are two packages that can be used to parse <code>mzIdentML</code> files, namely <code>mzR</code> (that we have already used for raw data) and <code>mzID</code>. The major difference is that the former leverages C++ code from <code>proteowizard</code> and is hence faster than the latter (which uses the <code>XML</code> R package). They both work in similar ways.</p>
<pre><code>|Data type      |File format |Data structure |Package |
|:--------------|:-----------|:--------------|:-------|
|Identification |mzIdentML   |mzRident       |mzR     |
|Identification |mzIdentML   |mzID           |mzID    |</code></pre>
<p>Which of these packages is used by <code>PSM()</code> can be defined by the <code>parser</code> argument, as documented in <code>?PSM</code>.</p>
<section id="mzid" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="mzid"><code>mzID</code></h3>
<p>The main functions are <code>mzID</code> to read the data into a dedicated data class and <code>flatten</code> to transform it into a <code>data.frame</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">mzID</span><span class="op">)</span></span>
<span><span class="va">idf</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">ident</a></span><span class="op">(</span>full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mzID/man/mzID.html">mzID</a></span><span class="op">(</span><span class="va">idf</span><span class="op">)</span></span>
<span><span class="co">##  reading TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid... DONE!</span></span>
<span><span class="va">id</span></span>
<span><span class="co">##  An mzID object</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Software used:   MS-GF+ (version: Beta (v10072))</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Rawfile:         /home/lg390/dev/01_svn/workflows/proteomics/TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Database:        /home/lg390/dev/01_svn/workflows/proteomics/erwinia_carotovora.fasta</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Number of scans: 5343</span></span>
<span><span class="co">##  Number of PSM's: 5656</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Various data can be extracted from the <code>mzID</code> object, using one of the accessor functions such as <code>database</code>, <code>software</code>, <code>scans</code>, <code>peptides</code>, … The object can also be converted into a <code>data.frame</code> using the <code>flatten</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mzID/man/flatten.html">flatten</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##                                       spectrumid scan number(s)</span></span>
<span><span class="co">##  1 controllerType=0 controllerNumber=1 scan=5782           5782</span></span>
<span><span class="co">##  2 controllerType=0 controllerNumber=1 scan=6037           6037</span></span>
<span><span class="co">##  3 controllerType=0 controllerNumber=1 scan=5235           5235</span></span>
<span><span class="co">##  4 controllerType=0 controllerNumber=1 scan=5397           5397</span></span>
<span><span class="co">##  5 controllerType=0 controllerNumber=1 scan=6075           6075</span></span>
<span><span class="co">##  6 controllerType=0 controllerNumber=1 scan=5761           5761</span></span>
<span><span class="co">##    acquisitionnum passthreshold rank calculatedmasstocharge</span></span>
<span><span class="co">##  1           5782          TRUE    1              1080.2321</span></span>
<span><span class="co">##  2           6037          TRUE    1              1002.2115</span></span>
<span><span class="co">##  3           5235          TRUE    1              1189.2800</span></span>
<span><span class="co">##  4           5397          TRUE    1               960.5365</span></span>
<span><span class="co">##  5           6075          TRUE    1              1264.3419</span></span>
<span><span class="co">##  6           5761          TRUE    1              1268.6501</span></span>
<span><span class="co">##    experimentalmasstocharge chargestate ms-gf:denovoscore ms-gf:evalue</span></span>
<span><span class="co">##  1                1080.2325           3               174 1.086033e-20</span></span>
<span><span class="co">##  2                1002.2089           3               245 1.988774e-19</span></span>
<span><span class="co">##  3                1189.2836           3               264 5.129649e-19</span></span>
<span><span class="co">##  4                 960.5365           3               178 5.163566e-18</span></span>
<span><span class="co">##  5                1264.3409           3               252 4.356914e-17</span></span>
<span><span class="co">##  6                1268.6429           2               138 4.658952e-17</span></span>
<span><span class="co">##    ms-gf:pepqvalue ms-gf:qvalue ms-gf:rawscore ms-gf:specevalue</span></span>
<span><span class="co">##  1               0            0            147     3.764831e-27</span></span>
<span><span class="co">##  2               0            0            214     6.902626e-26</span></span>
<span><span class="co">##  3               0            0            211     1.778789e-25</span></span>
<span><span class="co">##  4               0            0            154     1.792541e-24</span></span>
<span><span class="co">##  5               0            0            188     1.510364e-23</span></span>
<span><span class="co">##  6               0            0            123     1.618941e-23</span></span>
<span><span class="co">##    assumeddissociationmethod isotopeerror isdecoy post pre end start</span></span>
<span><span class="co">##  1                       HCD            0   FALSE    S   R  84    50</span></span>
<span><span class="co">##  2                       HCD            0   FALSE    R   K 315   288</span></span>
<span><span class="co">##  3                       HCD            0   FALSE    A   R 224   192</span></span>
<span><span class="co">##  4                       HCD            0   FALSE    -   R 290   264</span></span>
<span><span class="co">##  5                       HCD            0   FALSE    F   R 153   119</span></span>
<span><span class="co">##  6                       HCD            0   FALSE    Y   K 286   264</span></span>
<span><span class="co">##    accession length                                       description</span></span>
<span><span class="co">##  1   ECA1932    155                        outer membrane lipoprotein</span></span>
<span><span class="co">##  2   ECA1147    434                                    trigger factor</span></span>
<span><span class="co">##  3   ECA0013    295                ribose-binding periplasmic protein</span></span>
<span><span class="co">##  4   ECA1731    290                                         flagellin</span></span>
<span><span class="co">##  5   ECA1443    298      UTP--glucose-1-phosphate uridylyltransferase</span></span>
<span><span class="co">##  6   ECA1444    468 6-phosphogluconate dehydrogenase, decarboxylating</span></span>
<span><span class="co">##                                 pepseq modified modification</span></span>
<span><span class="co">##  1 PVQIQAGEDSNVIGALGGAVLGGFLGNTIGGGSGR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  2        TQVLDGLINANDIEVPVALIDGEIDVLR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  3   TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  4         SQILQQAGTSVLSQANQVPQTVLSLLR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  5 PIIGDNPFVVVLPDVVLDESTADQTQENLALLISR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##  6             WTSQSSLDLGEPLSLITESVFAR    FALSE         &lt;NA&gt;</span></span>
<span><span class="co">##                                                                 idFile</span></span>
<span><span class="co">##  1 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  2 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  3 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  4 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  5 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##  6 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid</span></span>
<span><span class="co">##                                                           spectrumFile</span></span>
<span><span class="co">##  1 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  2 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  3 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  4 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  5 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##  6 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</span></span>
<span><span class="co">##                databaseFile</span></span>
<span><span class="co">##  1 erwinia_carotovora.fasta</span></span>
<span><span class="co">##  2 erwinia_carotovora.fasta</span></span>
<span><span class="co">##  3 erwinia_carotovora.fasta</span></span>
<span><span class="co">##  4 erwinia_carotovora.fasta</span></span>
<span><span class="co">##  5 erwinia_carotovora.fasta</span></span>
<span><span class="co">##  6 erwinia_carotovora.fasta</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="mzr" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="mzr"><code>mzR</code></h3>
<p>The <code>mzR</code> interface provides a similar interface. It is however much faster as it does not read all the data into memory and only extracts relevant data on demand. It has also accessor functions such as <code>softwareInfo</code>, <code>mzidInfo</code>, … (use <code>showMethods(classes = "mzRident", where = "package:mzR")</code>) to see all available methods.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/mzR/">mzR</a></span><span class="op">)</span></span>
<span><span class="va">id2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mzR/man/openMSfile.html">openIDfile</a></span><span class="op">(</span><span class="va">idf</span><span class="op">)</span></span>
<span><span class="va">id2</span></span>
<span><span class="co">##  Identification file handle.</span></span>
<span><span class="co">##  Filename:  TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid </span></span>
<span><span class="co">##  Number of psms:  5759</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/mzR/man/metadata.html">softwareInfo</a></span><span class="op">(</span><span class="va">id2</span><span class="op">)</span></span>
<span><span class="co">##  [1] "MS-GF+ Beta (v10072) "                        </span></span>
<span><span class="co">##  [2] "ProteoWizard MzIdentML 3.0.21263 ProteoWizard"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The identification data can be accessed as a <code>data.frame</code> with the <code>psms</code> accessor.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html">psms</a></span><span class="op">(</span><span class="va">id2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##                                       spectrumID chargeState rank</span></span>
<span><span class="co">##  1 controllerType=0 controllerNumber=1 scan=5782           3    1</span></span>
<span><span class="co">##  2 controllerType=0 controllerNumber=1 scan=6037           3    1</span></span>
<span><span class="co">##  3 controllerType=0 controllerNumber=1 scan=5235           3    1</span></span>
<span><span class="co">##  4 controllerType=0 controllerNumber=1 scan=5397           3    1</span></span>
<span><span class="co">##  5 controllerType=0 controllerNumber=1 scan=6075           3    1</span></span>
<span><span class="co">##  6 controllerType=0 controllerNumber=1 scan=5761           2    1</span></span>
<span><span class="co">##    passThreshold experimentalMassToCharge calculatedMassToCharge</span></span>
<span><span class="co">##  1          TRUE                1080.2325              1080.2321</span></span>
<span><span class="co">##  2          TRUE                1002.2089              1002.2115</span></span>
<span><span class="co">##  3          TRUE                1189.2836              1189.2800</span></span>
<span><span class="co">##  4          TRUE                 960.5365               960.5365</span></span>
<span><span class="co">##  5          TRUE                1264.3409              1264.3419</span></span>
<span><span class="co">##  6          TRUE                1268.6429              1268.6501</span></span>
<span><span class="co">##                               sequence peptideRef modNum isDecoy post pre</span></span>
<span><span class="co">##  1 PVQIQAGEDSNVIGALGGAVLGGFLGNTIGGGSGR       Pep1      0   FALSE    S   R</span></span>
<span><span class="co">##  2        TQVLDGLINANDIEVPVALIDGEIDVLR       Pep2      0   FALSE    R   K</span></span>
<span><span class="co">##  3   TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR       Pep3      0   FALSE    A   R</span></span>
<span><span class="co">##  4         SQILQQAGTSVLSQANQVPQTVLSLLR       Pep4      0   FALSE    -   R</span></span>
<span><span class="co">##  5 PIIGDNPFVVVLPDVVLDESTADQTQENLALLISR       Pep5      0   FALSE    F   R</span></span>
<span><span class="co">##  6             WTSQSSLDLGEPLSLITESVFAR       Pep6      0   FALSE    Y   K</span></span>
<span><span class="co">##    start end DatabaseAccess DBseqLength DatabaseSeq</span></span>
<span><span class="co">##  1    50  84        ECA1932         155            </span></span>
<span><span class="co">##  2   288 315        ECA1147         434            </span></span>
<span><span class="co">##  3   192 224        ECA0013         295            </span></span>
<span><span class="co">##  4   264 290        ECA1731         290            </span></span>
<span><span class="co">##  5   119 153        ECA1443         298            </span></span>
<span><span class="co">##  6   264 286        ECA1444         468            </span></span>
<span><span class="co">##                                          DatabaseDescription scan.number.s.</span></span>
<span><span class="co">##  1                        ECA1932 outer membrane lipoprotein           5782</span></span>
<span><span class="co">##  2                                    ECA1147 trigger factor           6037</span></span>
<span><span class="co">##  3                ECA0013 ribose-binding periplasmic protein           5235</span></span>
<span><span class="co">##  4                                         ECA1731 flagellin           5397</span></span>
<span><span class="co">##  5      ECA1443 UTP--glucose-1-phosphate uridylyltransferase           6075</span></span>
<span><span class="co">##  6 ECA1444 6-phosphogluconate dehydrogenase, decarboxylating           5761</span></span>
<span><span class="co">##    acquisitionNum</span></span>
<span><span class="co">##  1           5782</span></span>
<span><span class="co">##  2           6037</span></span>
<span><span class="co">##  3           5235</span></span>
<span><span class="co">##  4           5397</span></span>
<span><span class="co">##  5           6075</span></span>
<span><span class="co">##  6           5761</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/30-quant.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quantitative data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/99-si.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Additional materials and help</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="../LICENSE.html">License</a>
  </li>  
</ul>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>